<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.551">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Jimena Monzón Sandoval">
<meta name="dcterms.date" content="2024-04-05">

<title>Protein Protein Interaction (PPI) resources</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="ppi_network_files/libs/clipboard/clipboard.min.js"></script>
<script src="ppi_network_files/libs/quarto-html/quarto.js"></script>
<script src="ppi_network_files/libs/quarto-html/popper.min.js"></script>
<script src="ppi_network_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="ppi_network_files/libs/quarto-html/anchor.min.js"></script>
<link href="ppi_network_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="ppi_network_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="ppi_network_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="ppi_network_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="ppi_network_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#resources" id="toc-resources" class="nav-link active" data-scroll-target="#resources">Resources</a>
  <ul class="collapse">
  <li><a href="#sources" id="toc-sources" class="nav-link" data-scroll-target="#sources">Sources</a></li>
  <li><a href="#id-mapping" id="toc-id-mapping" class="nav-link" data-scroll-target="#id-mapping">ID Mapping</a>
  <ul class="collapse">
  <li><a href="#entrez" id="toc-entrez" class="nav-link" data-scroll-target="#entrez">Entrez</a></li>
  <li><a href="#uniprot" id="toc-uniprot" class="nav-link" data-scroll-target="#uniprot">Uniprot</a></li>
  <li><a href="#ensembl-protein-id" id="toc-ensembl-protein-id" class="nav-link" data-scroll-target="#ensembl-protein-id">Ensembl Protein ID</a></li>
  </ul></li>
  <li><a href="#apid" id="toc-apid" class="nav-link" data-scroll-target="#apid">APID</a></li>
  <li><a href="#biogrid" id="toc-biogrid" class="nav-link" data-scroll-target="#biogrid">BIOGRID</a></li>
  <li><a href="#bioplex" id="toc-bioplex" class="nav-link" data-scroll-target="#bioplex">BIOPLEX</a>
  <ul class="collapse">
  <li><a href="#t" id="toc-t" class="nav-link" data-scroll-target="#t">293T</a></li>
  <li><a href="#hct116" id="toc-hct116" class="nav-link" data-scroll-target="#hct116">HCT116</a></li>
  </ul></li>
  <li><a href="#corum" id="toc-corum" class="nav-link" data-scroll-target="#corum">CORUM</a>
  <ul class="collapse">
  <li><a href="#all-complexes" id="toc-all-complexes" class="nav-link" data-scroll-target="#all-complexes">All complexes</a></li>
  <li><a href="#human-complexes" id="toc-human-complexes" class="nav-link" data-scroll-target="#human-complexes">Human complexes</a></li>
  <li><a href="#partial-complexes" id="toc-partial-complexes" class="nav-link" data-scroll-target="#partial-complexes">Partial complexes</a></li>
  </ul></li>
  <li><a href="#hitpredict" id="toc-hitpredict" class="nav-link" data-scroll-target="#hitpredict">HITPREDICT</a></li>
  <li><a href="#huri" id="toc-huri" class="nav-link" data-scroll-target="#huri">HURI</a></li>
  <li><a href="#intact" id="toc-intact" class="nav-link" data-scroll-target="#intact">INTACT</a></li>
  <li><a href="#mint" id="toc-mint" class="nav-link" data-scroll-target="#mint">MINT</a></li>
  <li><a href="#reactome" id="toc-reactome" class="nav-link" data-scroll-target="#reactome">REACTOME</a></li>
  <li><a href="#string" id="toc-string" class="nav-link" data-scroll-target="#string">STRING</a>
  <ul class="collapse">
  <li><a href="#protein-links" id="toc-protein-links" class="nav-link" data-scroll-target="#protein-links">Protein links</a></li>
  <li><a href="#protein-physical" id="toc-protein-physical" class="nav-link" data-scroll-target="#protein-physical">Protein physical</a></li>
  </ul></li>
  <li><a href="#overview-resources" id="toc-overview-resources" class="nav-link" data-scroll-target="#overview-resources">Overview resources</a></li>
  <li><a href="#merged-ppi-network" id="toc-merged-ppi-network" class="nav-link" data-scroll-target="#merged-ppi-network">Merged PPI network</a>
  <ul class="collapse">
  <li><a href="#all" id="toc-all" class="nav-link" data-scroll-target="#all">All</a></li>
  </ul></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul></li>
  </ul>
</nav>
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Protein Protein Interaction (PPI) resources</h1>
</div>


<div class="quarto-title-meta-author">
  <div class="quarto-title-meta-heading">Author</div>
  <div class="quarto-title-meta-heading">Affiliation</div>
  
    <div class="quarto-title-meta-contents">
    <p class="author">Jimena Monzón Sandoval <a href="https://orcid.org/0000-0002-4190-2188" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
  </div>
  <div class="quarto-title-meta-contents">
        <p class="affiliation">
            <a href="www.cardiff.ac.uk">
            Cardiff University
            </a>
          </p>
      </div>
  </div>

<div class="quarto-title-meta">

      
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">April 5, 2024</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<section id="resources" class="level1">
<h1>Resources</h1>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(stringr)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(R.utils)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(org.Hs.eg.db)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ensembldb)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(EnsDb.Hsapiens.v86)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggrepel)</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(MetBrewer)</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(cowplot)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="sources" class="level2">
<h2 class="anchored" data-anchor-id="sources">Sources</h2>
<p>Links to each database:</p>
<ul>
<li><p><a href="http://cicblade.dep.usal.es:8080/APID/init.action"><em>APID</em></a> Agile Protein Interactomes DataServer. Integration of known experimentally validated protein-protein physical interactions (PPIs).</p></li>
<li><p><a href="https://thebiogrid.org/"><em>BIOGRID</em></a> Biological General Repository for Interaction Datasets. Genetic and protein interaction data from model organisms and humans.</p></li>
<li><p><a href="https://bioplex.hms.harvard.edu/"><em>BIOPLEX</em></a>. Bioplex 3.0, comprehensive experimentally derived model of the human interactome.</p></li>
<li><p><a href="http://mips.helmholtz-muenchen.de/corum/"><em>CORUM</em></a> the COmprehensive ResoUrce of Mammalian protein complexes.</p></li>
<li><p><a href="http://www.hitpredict.org/"><em>HITPREDICT</em></a> Physical protein-protein interactions with reliability scores.</p></li>
<li><p><a href="http://www.interactome-atlas.org/"><em>HURI</em></a> The Human Reference Interactome.</p></li>
<li><p><a href="https://www.ebi.ac.uk/intact/home"><em>INTACT</em></a> IntAct Molecular Interaction Database.</p></li>
<li><p><a href="https://mint.bio.uniroma2.it/"><em>MINT</em></a> The Molecular INTeraction Database.</p></li>
<li><p><a href="https://reactome.org/download-data"><em>REACTOME</em></a>. Open-source, open access, manually curated and peer-reviewed pathway database</p></li>
<li><p><a href="https://string-db.org/"><em>STRING</em></a>. STRING is a Core Data Resource as designated by Global Biodata Coalition and ELIXIR.</p></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="at">path =</span> <span class="st">"./02_Datasets/"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="id-mapping" class="level2">
<h2 class="anchored" data-anchor-id="id-mapping">ID Mapping</h2>
<ul>
<li>All datasets were mapped to Ensembl Gene ID from either Entrez, Uniprot or Ensembl Protein ID.</li>
</ul>
<section id="entrez" class="level3">
<h3 class="anchored" data-anchor-id="entrez">Entrez</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Annotations Entrez ID to Ensembl Gene ID</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> org.Hs.egENSEMBL</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Get the entrez gene IDs that are mapped to an Ensembl ID</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>mapped_genes <span class="ot">&lt;-</span> <span class="fu">mappedkeys</span>(x)</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert to a list</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>xx <span class="ot">&lt;-</span> <span class="fu">as.list</span>(x[mapped_genes])</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Built a Dataframe with unique Entrez IDs, possible multiple Ensembl Gene IDs delimited by "||"</span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>entrez_to_ensembl <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="cn">NA</span>, <span class="at">nrow =</span> <span class="dv">1</span>, <span class="at">ncol =</span> <span class="dv">2</span>)</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(xx)){</span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">length</span>(<span class="fu">unlist</span>(xx[i])) <span class="sc">==</span> <span class="dv">1</span>){</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>    gci <span class="ot">&lt;-</span> <span class="fu">cbind</span>(<span class="fu">names</span>(xx[i]), <span class="fu">unlist</span>(xx[i]))</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>    gci <span class="ot">&lt;-</span> <span class="fu">cbind</span>(<span class="fu">names</span>(xx[i]), <span class="fu">paste0</span>(<span class="fu">unlist</span>(xx[i]), <span class="at">collapse =</span> <span class="st">"||"</span>))</span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a>  entrez_to_ensembl <span class="ot">&lt;-</span> <span class="fu">rbind</span>(entrez_to_ensembl, gci)</span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rm</span>(gci)</span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(i, x, mapped_genes, xx)</span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a>entrez_to_ensembl <span class="ot">&lt;-</span> entrez_to_ensembl[<span class="sc">-</span><span class="dv">1</span>, ]</span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(entrez_to_ensembl) <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(entrez_to_ensembl)</span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(entrez_to_ensembl) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Entrez_ID"</span>, <span class="st">"Ensembl_GeneID"</span>)</span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a>entrez_to_ensembl <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(entrez_to_ensembl)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="uniprot" class="level3">
<h3 class="anchored" data-anchor-id="uniprot">Uniprot</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Uniprot annotations from Ensembl</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>edb <span class="ot">&lt;-</span> EnsDb.Hsapiens.v86</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>txs <span class="ot">&lt;-</span> <span class="fu">genes</span>(edb, <span class="at">columns =</span> <span class="fu">c</span>(<span class="st">"uniprot_id"</span>))</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>uniprot_to_ensembl <span class="ot">&lt;-</span> <span class="fu">mcols</span>(txs)</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(uniprot_to_ensembl) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Uniprot_ID"</span>, <span class="st">"Ensembl_GeneID"</span>)</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove genes that do not match any Uniprot ID</span></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">length</span>(<span class="fu">which</span>(<span class="fu">is.na</span>(uniprot_to_ensembl<span class="sc">$</span>Uniprot_ID) <span class="sc">==</span> <span class="cn">TRUE</span>)) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a> uniprot_to_ensembl <span class="ot">&lt;-</span> uniprot_to_ensembl[<span class="sc">-</span><span class="fu">which</span>(<span class="fu">is.na</span>(uniprot_to_ensembl<span class="sc">$</span>Uniprot_ID) <span class="sc">==</span> <span class="cn">TRUE</span>), ] </span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(edb, txs)</span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>uniprot_to_ensembl <span class="ot">&lt;-</span> uniprot_to_ensembl[<span class="fu">order</span>(uniprot_to_ensembl<span class="sc">$</span>Uniprot_ID), ]</span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a>uniprot_to_ensembl_unique <span class="ot">&lt;-</span> <span class="fu">aggregate.data.frame</span>(<span class="at">x =</span> uniprot_to_ensembl,</span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a>                                                  <span class="at">by =</span> <span class="fu">list</span>(uniprot_to_ensembl<span class="sc">$</span>Uniprot_ID),</span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a>                                                  <span class="at">FUN =</span> <span class="cf">function</span>(x) <span class="fu">paste0</span>(x, <span class="at">collapse =</span> <span class="st">"||"</span>))</span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(uniprot_to_ensembl)</span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a>uniprot_to_ensembl <span class="ot">&lt;-</span> uniprot_to_ensembl_unique[, <span class="fu">c</span>(<span class="st">"Group.1"</span>, <span class="st">"Ensembl_GeneID"</span>)]</span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(uniprot_to_ensembl) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Uniprot_ID"</span>, <span class="st">"Ensembl_GeneID"</span>)</span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(uniprot_to_ensembl_unique)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="ensembl-protein-id" class="level3">
<h3 class="anchored" data-anchor-id="ensembl-protein-id">Ensembl Protein ID</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Uniprot annotations from Ensembl</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>edb <span class="ot">&lt;-</span> EnsDb.Hsapiens.v86</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>txs <span class="ot">&lt;-</span> <span class="fu">genes</span>(edb, <span class="at">columns =</span> <span class="fu">c</span>(<span class="st">"protein_id"</span>))</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>proteinID_to_ensembl <span class="ot">&lt;-</span> <span class="fu">mcols</span>(txs)</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(proteinID_to_ensembl) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"protein_id"</span>, <span class="st">"Ensembl_GeneID"</span>)</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove genes that do not match any protein ID</span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">length</span>(<span class="fu">which</span>(<span class="fu">is.na</span>(proteinID_to_ensembl<span class="sc">$</span>protein_id) <span class="sc">==</span> <span class="cn">TRUE</span>)) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a> proteinID_to_ensembl <span class="ot">&lt;-</span> proteinID_to_ensembl[<span class="sc">-</span><span class="fu">which</span>(<span class="fu">is.na</span>(proteinID_to_ensembl<span class="sc">$</span>protein_id) <span class="sc">==</span> <span class="cn">TRUE</span>), ] </span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(edb, txs)</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>proteinID_to_ensembl <span class="ot">&lt;-</span> proteinID_to_ensembl[<span class="fu">order</span>(proteinID_to_ensembl<span class="sc">$</span>protein_id), ]</span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a>proteinID_to_ensembl_unique <span class="ot">&lt;-</span> <span class="fu">aggregate.data.frame</span>(<span class="at">x =</span> proteinID_to_ensembl,</span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a>                                                  <span class="at">by =</span> <span class="fu">list</span>(proteinID_to_ensembl<span class="sc">$</span>protein_id),</span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a>                                                  <span class="at">FUN =</span> <span class="cf">function</span>(x) <span class="fu">paste0</span>(x, <span class="at">collapse =</span> <span class="st">"||"</span>))</span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(proteinID_to_ensembl)</span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a>proteinID_to_ensembl <span class="ot">&lt;-</span> proteinID_to_ensembl_unique[, <span class="fu">c</span>(<span class="st">"Group.1"</span>, <span class="st">"Ensembl_GeneID"</span>)]</span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(proteinID_to_ensembl) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Protein_ID"</span>, <span class="st">"Ensembl_GeneID"</span>)</span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(proteinID_to_ensembl_unique)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="apid" class="level2">
<h2 class="anchored" data-anchor-id="apid">APID</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># User friendly download through</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="co"># http://cicblade.dep.usal.es:8080/APID/init.action</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Organism Homo Sapiens</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Level 0. all reported proteins pairs</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="co"># filter out inter-species interactions select "YES"</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a><span class="co"># File name 9606_noISI_Q3.txt</span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="at">path =</span> <span class="st">"./02_Datasets/APID/"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Apid #</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>db <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(<span class="st">"./02_Datasets/APID/9606_noISI_Q3.txt"</span>, <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, </span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>                 <span class="at">quote =</span> <span class="cn">NULL</span>, <span class="at">header =</span> <span class="cn">TRUE</span>, <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>)</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Mapping through uniprot</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>db <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(db, </span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>                 <span class="at">Ensembl_Gene_Interactor_A =</span> uniprot_to_ensembl<span class="sc">$</span>Ensembl_GeneID[<span class="fu">match</span>(<span class="at">x =</span> db<span class="sc">$</span>UniprotID_A, <span class="at">table =</span> uniprot_to_ensembl<span class="sc">$</span>Uniprot_ID)],</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>                 <span class="at">Ensembl_Gene_Interactor_B =</span> uniprot_to_ensembl<span class="sc">$</span>Ensembl_GeneID[<span class="fu">match</span>(<span class="at">x =</span> db<span class="sc">$</span>UniprotID_B, <span class="at">table =</span> uniprot_to_ensembl<span class="sc">$</span>Uniprot_ID)])</span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove interactions where either interactor cannot be mapped to Ensembl Gene IDs</span></span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>nm <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">c</span>(<span class="fu">which</span>(<span class="fu">is.na</span>(db<span class="sc">$</span>Ensembl_Gene_Interactor_A) <span class="sc">==</span> <span class="cn">TRUE</span>),</span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a>               <span class="fu">which</span>(<span class="fu">is.na</span>(db<span class="sc">$</span>Ensembl_Gene_Interactor_B) <span class="sc">==</span> <span class="cn">TRUE</span>)))</span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">length</span>(nm)<span class="sc">&gt;</span> <span class="dv">0</span>){ db <span class="ot">&lt;-</span> db[<span class="sc">-</span>nm, ]}</span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(nm)</span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Separate into interactions with a single and multiple Ensembl Gene IDs</span></span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true" tabindex="-1"></a>db_m2m <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">c</span>(<span class="fu">which</span>(<span class="fu">str_detect</span>(<span class="at">string =</span> db<span class="sc">$</span>Ensembl_Gene_Interactor_A, <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">|"</span>) <span class="sc">==</span> <span class="cn">TRUE</span>),</span>
<span id="cb7-18"><a href="#cb7-18" aria-hidden="true" tabindex="-1"></a>                   <span class="fu">which</span>(<span class="fu">str_detect</span>(<span class="at">string =</span> db<span class="sc">$</span>Ensembl_Gene_Interactor_B, <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">|"</span>) <span class="sc">==</span> <span class="cn">TRUE</span>)))</span>
<span id="cb7-19"><a href="#cb7-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-20"><a href="#cb7-20" aria-hidden="true" tabindex="-1"></a>db_o2o <span class="ot">&lt;-</span> <span class="fu">setdiff</span>(<span class="at">x =</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(db), <span class="at">y =</span> db_m2m)</span>
<span id="cb7-21"><a href="#cb7-21" aria-hidden="true" tabindex="-1"></a>db_m2m <span class="ot">&lt;-</span> db[db_m2m, ]</span>
<span id="cb7-22"><a href="#cb7-22" aria-hidden="true" tabindex="-1"></a>db_o2o <span class="ot">&lt;-</span> db[db_o2o, ]</span>
<span id="cb7-23"><a href="#cb7-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-24"><a href="#cb7-24" aria-hidden="true" tabindex="-1"></a><span class="co"># Expand interactions with multiple ensembl gene IDs</span></span>
<span id="cb7-25"><a href="#cb7-25" aria-hidden="true" tabindex="-1"></a>output <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="cn">NA</span>, <span class="at">nrow =</span> <span class="dv">1</span>, <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">dimnames =</span> <span class="fu">list</span>(<span class="st">"ToRemove"</span>, <span class="fu">c</span>(<span class="st">"Ensembl_Gene_Interactor_A"</span>, <span class="st">"Ensembl_Gene_Interactor_B"</span>)))</span>
<span id="cb7-26"><a href="#cb7-26" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(db_m2m))){</span>
<span id="cb7-27"><a href="#cb7-27" aria-hidden="true" tabindex="-1"></a>  in_a <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">str_split</span>(<span class="at">string =</span> db_m2m<span class="sc">$</span>Ensembl_Gene_Interactor_A[i], <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">|</span><span class="sc">\\</span><span class="st">|"</span>))</span>
<span id="cb7-28"><a href="#cb7-28" aria-hidden="true" tabindex="-1"></a>  in_b <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">str_split</span>(<span class="at">string =</span> db_m2m<span class="sc">$</span>Ensembl_Gene_Interactor_B[i], <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">|</span><span class="sc">\\</span><span class="st">|"</span>))</span>
<span id="cb7-29"><a href="#cb7-29" aria-hidden="true" tabindex="-1"></a>  ab <span class="ot">&lt;-</span> <span class="fu">expand.grid</span>(<span class="st">"Ensembl_Gene_Interactor_A"</span> <span class="ot">=</span> in_a, <span class="st">"Ensembl_Gene_Interactor_B"</span> <span class="ot">=</span> in_b)</span>
<span id="cb7-30"><a href="#cb7-30" aria-hidden="true" tabindex="-1"></a>  output <span class="ot">&lt;-</span> <span class="fu">rbind</span>(output, ab)</span>
<span id="cb7-31"><a href="#cb7-31" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb7-32"><a href="#cb7-32" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(in_a, in_b, ab, i)</span>
<span id="cb7-33"><a href="#cb7-33" aria-hidden="true" tabindex="-1"></a>output <span class="ot">&lt;-</span> output[<span class="sc">-</span><span class="dv">1</span>, ]</span>
<span id="cb7-34"><a href="#cb7-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-35"><a href="#cb7-35" aria-hidden="true" tabindex="-1"></a><span class="co"># Combine interactions that map a single and multiple Ensembl Gene IDs </span></span>
<span id="cb7-36"><a href="#cb7-36" aria-hidden="true" tabindex="-1"></a>db_ens <span class="ot">&lt;-</span> <span class="fu">rbind</span>(db_o2o[, <span class="fu">c</span>(<span class="st">"Ensembl_Gene_Interactor_A"</span>, <span class="st">"Ensembl_Gene_Interactor_B"</span>)], </span>
<span id="cb7-37"><a href="#cb7-37" aria-hidden="true" tabindex="-1"></a>                output)</span>
<span id="cb7-38"><a href="#cb7-38" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(db, db_m2m, db_o2o, output)</span>
<span id="cb7-39"><a href="#cb7-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-40"><a href="#cb7-40" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove self interactions</span></span>
<span id="cb7-41"><a href="#cb7-41" aria-hidden="true" tabindex="-1"></a>mark_self <span class="ot">&lt;-</span> <span class="fu">apply</span>(db_ens, <span class="dv">1</span>, <span class="at">FUN =</span> <span class="cf">function</span>(x) <span class="fu">length</span>(<span class="fu">unique</span>(x)))</span>
<span id="cb7-42"><a href="#cb7-42" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">length</span>(<span class="fu">which</span>(mark_self <span class="sc">==</span> <span class="dv">1</span>))<span class="sc">&gt;</span> <span class="dv">0</span>){db_ens <span class="ot">&lt;-</span> db_ens[<span class="sc">-</span><span class="fu">which</span>(mark_self <span class="sc">==</span> <span class="dv">1</span>), ]}</span>
<span id="cb7-43"><a href="#cb7-43" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(mark_self)</span>
<span id="cb7-44"><a href="#cb7-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-45"><a href="#cb7-45" aria-hidden="true" tabindex="-1"></a><span class="co"># Re-order interactions to remove duplicates</span></span>
<span id="cb7-46"><a href="#cb7-46" aria-hidden="true" tabindex="-1"></a>db_ens_order <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">apply</span>(db_ens, <span class="dv">1</span>, <span class="at">FUN =</span> <span class="cf">function</span>(x) x[<span class="fu">order</span>(x)]))</span>
<span id="cb7-47"><a href="#cb7-47" aria-hidden="true" tabindex="-1"></a>db_ens <span class="ot">&lt;-</span> <span class="fu">unique</span>(db_ens_order) </span>
<span id="cb7-48"><a href="#cb7-48" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(db_ens_order)</span>
<span id="cb7-49"><a href="#cb7-49" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-50"><a href="#cb7-50" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert to data frame</span></span>
<span id="cb7-51"><a href="#cb7-51" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(db_ens) <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(db_ens)</span>
<span id="cb7-52"><a href="#cb7-52" aria-hidden="true" tabindex="-1"></a>db_ens <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(db_ens)</span>
<span id="cb7-53"><a href="#cb7-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-54"><a href="#cb7-54" aria-hidden="true" tabindex="-1"></a><span class="co"># Save as tab delimited file </span></span>
<span id="cb7-55"><a href="#cb7-55" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(<span class="at">x =</span> db_ens, </span>
<span id="cb7-56"><a href="#cb7-56" aria-hidden="true" tabindex="-1"></a>            <span class="at">file =</span> <span class="st">"./02_Datasets/APID/9606_noISI_Q3_EnsemblGeneID.txt"</span>, </span>
<span id="cb7-57"><a href="#cb7-57" aria-hidden="true" tabindex="-1"></a>            <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>, <span class="at">col.names =</span> <span class="cn">FALSE</span>, <span class="at">quote =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="biogrid" class="level2">
<h2 class="anchored" data-anchor-id="biogrid">BIOGRID</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Download data from Biogrid by species</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="at">path =</span> <span class="st">"./02_Datasets/BIOGRID/"</span>)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(<span class="at">url=</span> <span class="st">"https://downloads.thebiogrid.org/Download/BioGRID/Release-Archive/BIOGRID-4.4.230/BIOGRID-ORGANISM-4.4.230.tab3.zip"</span>,</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>              <span class="at">destfile =</span> <span class="st">"./02_Datasets/BIOGRID/BIOGRID-ORGANISM-4.4.230.tab3.zip"</span>) </span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="fu">unzip</span>(<span class="at">zipfile =</span> <span class="st">"./02_Datasets/BIOGRID/BIOGRID-ORGANISM-4.4.230.tab3.zip"</span>, </span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>      <span class="at">exdir =</span> <span class="st">"./02_Datasets/BIOGRID/BIOGRID-ORGANISM-4.4.230.tab3"</span>)</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a><span class="fu">unlink</span>(<span class="at">x =</span> <span class="st">"./02_Datasets/BIOGRID/BIOGRID-ORGANISM-4.4.230.tab3.zip/"</span>, <span class="at">recursive =</span> <span class="cn">TRUE</span>)</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Keep only Homo sapiens </span></span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>fn <span class="ot">&lt;-</span> <span class="fu">list.files</span>(<span class="at">path =</span> <span class="st">"./02_Datasets/BIOGRID/BIOGRID-ORGANISM-4.4.230.tab3/"</span>, <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>fn_hs <span class="ot">&lt;-</span> fn[<span class="fu">str_detect</span>(<span class="at">string =</span> fn, <span class="at">pattern =</span> <span class="st">"Homo_sapiens"</span>)]</span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>fn_rm <span class="ot">&lt;-</span> fn[<span class="fu">str_detect</span>(<span class="at">string =</span> fn, <span class="at">pattern =</span> <span class="st">"Homo_sapiens"</span>, <span class="at">negate =</span> <span class="cn">TRUE</span>)]</span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a><span class="fu">unlink</span>(<span class="at">x =</span> fn_rm)</span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(fn, fn_rm)</span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a><span class="co"># BIOGRID files to system variables </span></span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a><span class="fu">Sys.setenv</span>(<span class="at">BIOGRID_HS =</span> fn_hs)</span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a><span class="fu">Sys.setenv</span>(<span class="at">BIOGRID_HS_4R =</span> <span class="fu">paste0</span>(fn_hs, <span class="st">".4R"</span>, <span class="at">sep =</span> <span class="st">""</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Parse file to remove ' and # from file</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="fu">sed</span> <span class="st">"s/'//g"</span> <span class="va">$BIOGRID_HS</span> <span class="op">&gt;</span> <span class="va">$BIOGRID_HS_4R</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="fu">sed</span> <span class="st">"s/#//g"</span> <span class="va">$BIOGRID_HS_4R</span> <span class="op">&gt;</span> <span class="va">$BIOGRID_HS</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove intermediate file</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="fu">unlink</span>(<span class="at">x =</span> <span class="fu">paste0</span>(fn_hs, <span class="st">".4R"</span>, <span class="at">sep =</span> <span class="st">""</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Homo sapiens interactions from BIOGRID</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>db <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="at">file =</span> fn_hs, <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, </span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>                 <span class="at">header =</span> <span class="cn">TRUE</span>, <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>)</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(fn_hs)</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Keep only physical interactions between Homo sapiens  </span></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>db <span class="ot">&lt;-</span> db[<span class="fu">which</span>(db<span class="sc">$</span>Experimental.System.Type <span class="sc">==</span> <span class="st">"physical"</span>), ]</span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>db <span class="ot">&lt;-</span> db[<span class="fu">which</span>(db<span class="sc">$</span>Organism.Name.Interactor.A <span class="sc">==</span> <span class="st">"Homo sapiens"</span>), ]</span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>db <span class="ot">&lt;-</span> db[<span class="fu">which</span>(db<span class="sc">$</span>Organism.Name.Interactor.B <span class="sc">==</span> <span class="st">"Homo sapiens"</span>), ]</span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Interactions from Entrez to Ensembl Gene IDs</span></span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a>db <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(db, </span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a>                 <span class="at">Ensembl_Gene_Interactor_A =</span> entrez_to_ensembl<span class="sc">$</span>Ensembl_GeneID[<span class="fu">match</span>(<span class="at">x =</span> db<span class="sc">$</span>Entrez.Gene.Interactor.A, <span class="at">table =</span> entrez_to_ensembl<span class="sc">$</span>Entrez_ID)], </span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true" tabindex="-1"></a>                 <span class="at">Ensembl_Gene_Interactor_B =</span> entrez_to_ensembl<span class="sc">$</span>Ensembl_GeneID[<span class="fu">match</span>(<span class="at">x =</span> db<span class="sc">$</span>Entrez.Gene.Interactor.B, <span class="at">table =</span> entrez_to_ensembl<span class="sc">$</span>Entrez_ID)])</span>
<span id="cb11-15"><a href="#cb11-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-16"><a href="#cb11-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove interactions where either interactor cannot be mapped to Ensembl Gene IDs</span></span>
<span id="cb11-17"><a href="#cb11-17" aria-hidden="true" tabindex="-1"></a>nm <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">c</span>(<span class="fu">which</span>(<span class="fu">is.na</span>(db<span class="sc">$</span>Ensembl_Gene_Interactor_A) <span class="sc">==</span> <span class="cn">TRUE</span>),</span>
<span id="cb11-18"><a href="#cb11-18" aria-hidden="true" tabindex="-1"></a>               <span class="fu">which</span>(<span class="fu">is.na</span>(db<span class="sc">$</span>Ensembl_Gene_Interactor_B) <span class="sc">==</span> <span class="cn">TRUE</span>)))</span>
<span id="cb11-19"><a href="#cb11-19" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">length</span>(nm)<span class="sc">&gt;</span> <span class="dv">0</span>){ db <span class="ot">&lt;-</span> db[<span class="sc">-</span>nm, ]}</span>
<span id="cb11-20"><a href="#cb11-20" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(nm)</span>
<span id="cb11-21"><a href="#cb11-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-22"><a href="#cb11-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Separate into interactions with a single and multiple Ensembl Gene IDs</span></span>
<span id="cb11-23"><a href="#cb11-23" aria-hidden="true" tabindex="-1"></a>db_m2m <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">c</span>(<span class="fu">which</span>(<span class="fu">str_detect</span>(<span class="at">string =</span> db<span class="sc">$</span>Ensembl_Gene_Interactor_A, <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">|"</span>) <span class="sc">==</span> <span class="cn">TRUE</span>),</span>
<span id="cb11-24"><a href="#cb11-24" aria-hidden="true" tabindex="-1"></a>                   <span class="fu">which</span>(<span class="fu">str_detect</span>(<span class="at">string =</span> db<span class="sc">$</span>Ensembl_Gene_Interactor_B, <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">|"</span>) <span class="sc">==</span> <span class="cn">TRUE</span>)))</span>
<span id="cb11-25"><a href="#cb11-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-26"><a href="#cb11-26" aria-hidden="true" tabindex="-1"></a>db_o2o <span class="ot">&lt;-</span> <span class="fu">setdiff</span>(<span class="at">x =</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(db), <span class="at">y =</span> db_m2m)</span>
<span id="cb11-27"><a href="#cb11-27" aria-hidden="true" tabindex="-1"></a>db_m2m <span class="ot">&lt;-</span> db[db_m2m, ]</span>
<span id="cb11-28"><a href="#cb11-28" aria-hidden="true" tabindex="-1"></a>db_o2o <span class="ot">&lt;-</span> db[db_o2o, ]</span>
<span id="cb11-29"><a href="#cb11-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-30"><a href="#cb11-30" aria-hidden="true" tabindex="-1"></a><span class="co"># Expand interactions with multiple ensembl gene IDs</span></span>
<span id="cb11-31"><a href="#cb11-31" aria-hidden="true" tabindex="-1"></a>output <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="cn">NA</span>, <span class="at">nrow =</span> <span class="dv">1</span>, <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">dimnames =</span> <span class="fu">list</span>(<span class="st">"ToRemove"</span>, <span class="fu">c</span>(<span class="st">"Ensembl_Gene_Interactor_A"</span>, <span class="st">"Ensembl_Gene_Interactor_B"</span>)))</span>
<span id="cb11-32"><a href="#cb11-32" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(db_m2m))){</span>
<span id="cb11-33"><a href="#cb11-33" aria-hidden="true" tabindex="-1"></a>  in_a <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">str_split</span>(<span class="at">string =</span> db_m2m<span class="sc">$</span>Ensembl_Gene_Interactor_A[i], <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">|</span><span class="sc">\\</span><span class="st">|"</span>))</span>
<span id="cb11-34"><a href="#cb11-34" aria-hidden="true" tabindex="-1"></a>  in_b <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">str_split</span>(<span class="at">string =</span> db_m2m<span class="sc">$</span>Ensembl_Gene_Interactor_B[i], <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">|</span><span class="sc">\\</span><span class="st">|"</span>))</span>
<span id="cb11-35"><a href="#cb11-35" aria-hidden="true" tabindex="-1"></a>  ab <span class="ot">&lt;-</span> <span class="fu">expand.grid</span>(<span class="st">"Ensembl_Gene_Interactor_A"</span> <span class="ot">=</span> in_a, <span class="st">"Ensembl_Gene_Interactor_B"</span> <span class="ot">=</span> in_b)</span>
<span id="cb11-36"><a href="#cb11-36" aria-hidden="true" tabindex="-1"></a>  output <span class="ot">&lt;-</span> <span class="fu">rbind</span>(output, ab)</span>
<span id="cb11-37"><a href="#cb11-37" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb11-38"><a href="#cb11-38" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb11-39"><a href="#cb11-39" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(in_a, in_b, ab, i)</span>
<span id="cb11-40"><a href="#cb11-40" aria-hidden="true" tabindex="-1"></a>output <span class="ot">&lt;-</span> output[<span class="sc">-</span><span class="dv">1</span>, ]</span>
<span id="cb11-41"><a href="#cb11-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-42"><a href="#cb11-42" aria-hidden="true" tabindex="-1"></a><span class="co"># Combine interactions that map a single and multiple Ensembl Gene IDs </span></span>
<span id="cb11-43"><a href="#cb11-43" aria-hidden="true" tabindex="-1"></a>db_ens <span class="ot">&lt;-</span> <span class="fu">rbind</span>(db_o2o[, <span class="fu">c</span>(<span class="st">"Ensembl_Gene_Interactor_A"</span>, <span class="st">"Ensembl_Gene_Interactor_B"</span>)], </span>
<span id="cb11-44"><a href="#cb11-44" aria-hidden="true" tabindex="-1"></a>                output)</span>
<span id="cb11-45"><a href="#cb11-45" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(db, db_m2m, db_o2o, output)</span>
<span id="cb11-46"><a href="#cb11-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-47"><a href="#cb11-47" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove self interactions</span></span>
<span id="cb11-48"><a href="#cb11-48" aria-hidden="true" tabindex="-1"></a>mark_self <span class="ot">&lt;-</span> <span class="fu">apply</span>(db_ens, <span class="dv">1</span>, <span class="at">FUN =</span> <span class="cf">function</span>(x) <span class="fu">length</span>(<span class="fu">unique</span>(x)))</span>
<span id="cb11-49"><a href="#cb11-49" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">length</span>(<span class="fu">which</span>(mark_self <span class="sc">==</span> <span class="dv">1</span>))<span class="sc">&gt;</span> <span class="dv">0</span>){db_ens <span class="ot">&lt;-</span> db_ens[<span class="sc">-</span><span class="fu">which</span>(mark_self <span class="sc">==</span> <span class="dv">1</span>), ]}</span>
<span id="cb11-50"><a href="#cb11-50" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(mark_self)</span>
<span id="cb11-51"><a href="#cb11-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-52"><a href="#cb11-52" aria-hidden="true" tabindex="-1"></a><span class="co"># Re-order interactions to remove duplicates</span></span>
<span id="cb11-53"><a href="#cb11-53" aria-hidden="true" tabindex="-1"></a>db_ens_order <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">apply</span>(db_ens, <span class="dv">1</span>, <span class="at">FUN =</span> <span class="cf">function</span>(x) x[<span class="fu">order</span>(x)]))</span>
<span id="cb11-54"><a href="#cb11-54" aria-hidden="true" tabindex="-1"></a>db_ens <span class="ot">&lt;-</span> <span class="fu">unique</span>(db_ens_order) </span>
<span id="cb11-55"><a href="#cb11-55" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(db_ens_order)</span>
<span id="cb11-56"><a href="#cb11-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-57"><a href="#cb11-57" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert to data frame</span></span>
<span id="cb11-58"><a href="#cb11-58" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(db_ens) <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(db_ens)</span>
<span id="cb11-59"><a href="#cb11-59" aria-hidden="true" tabindex="-1"></a>db_ens <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(db_ens)</span>
<span id="cb11-60"><a href="#cb11-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-61"><a href="#cb11-61" aria-hidden="true" tabindex="-1"></a><span class="co"># Save as tab delimited file </span></span>
<span id="cb11-62"><a href="#cb11-62" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(<span class="at">x =</span> db_ens, </span>
<span id="cb11-63"><a href="#cb11-63" aria-hidden="true" tabindex="-1"></a>            <span class="at">file =</span> <span class="st">"./02_Datasets/BIOGRID/BIOGRID-ORGANISM-4.4.230.tab3_EnsemblGeneID.txt"</span>, </span>
<span id="cb11-64"><a href="#cb11-64" aria-hidden="true" tabindex="-1"></a>            <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>, <span class="at">col.names =</span> <span class="cn">FALSE</span>, <span class="at">quote =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="bioplex" class="level2">
<h2 class="anchored" data-anchor-id="bioplex">BIOPLEX</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Data download</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="at">path =</span> <span class="st">"./02_Datasets/BIOPLEX/"</span>)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Bioplex version 3.0 # </span></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Same name different file size for HCT116</span></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(<span class="at">url =</span> <span class="st">"https://bioplex.hms.harvard.edu/data/BioPlex_293T_Network_10K_Dec_2019.tsv"</span>,</span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>                            <span class="at">destfile =</span> <span class="st">"./02_Datasets/BIOPLEX/BioPlex_293T_Network_10K_Dec_2019.tsv"</span>)</span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(<span class="at">url =</span> <span class="st">"https://bioplex.hms.harvard.edu/data/BioPlex_HCT116_Network_5.5K_Dec_2019.tsv"</span>,</span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>              <span class="at">destfile =</span> <span class="st">"./02_Datasets/BIOPLEX/BioPlex_HCT116_Network_5.5K_Dec_2019.tsv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="t" class="level3">
<h3 class="anchored" data-anchor-id="t">293T</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Bioplex</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>db <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="at">file =</span> <span class="st">"./02_Datasets/BIOPLEX/BioPlex_293T_Network_10K_Dec_2019.tsv"</span>, <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">header =</span> <span class="cn">TRUE</span>)</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove isoform info to mapp Uniprot ID</span></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>db<span class="sc">$</span>UniprotA <span class="ot">&lt;-</span> <span class="fu">sapply</span>(<span class="fu">str_split</span>(<span class="at">string =</span> db<span class="sc">$</span>UniprotA, <span class="at">pattern =</span> <span class="st">"-"</span>), <span class="st">"["</span>, <span class="dv">1</span>)</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>db<span class="sc">$</span>UniprotB <span class="ot">&lt;-</span> <span class="fu">sapply</span>(<span class="fu">str_split</span>(<span class="at">string =</span> db<span class="sc">$</span>UniprotB, <span class="at">pattern =</span> <span class="st">"-"</span>), <span class="st">"["</span>, <span class="dv">1</span>)</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Mapping through uniprot</span></span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>db <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(db, </span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>                 <span class="at">Ensembl_Gene_Interactor_A =</span> uniprot_to_ensembl<span class="sc">$</span>Ensembl_GeneID[<span class="fu">match</span>(<span class="at">x =</span> db<span class="sc">$</span>UniprotA, <span class="at">table =</span> uniprot_to_ensembl<span class="sc">$</span>Uniprot_ID)],</span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>                 <span class="at">Ensembl_Gene_Interactor_B =</span> uniprot_to_ensembl<span class="sc">$</span>Ensembl_GeneID[<span class="fu">match</span>(<span class="at">x =</span> db<span class="sc">$</span>UniprotB, <span class="at">table =</span> uniprot_to_ensembl<span class="sc">$</span>Uniprot_ID)])</span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove interactions where either interactor cannot be mapped to Ensembl Gene IDs</span></span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a>nm <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">c</span>(<span class="fu">which</span>(<span class="fu">is.na</span>(db<span class="sc">$</span>Ensembl_Gene_Interactor_A) <span class="sc">==</span> <span class="cn">TRUE</span>),</span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a>               <span class="fu">which</span>(<span class="fu">is.na</span>(db<span class="sc">$</span>Ensembl_Gene_Interactor_B) <span class="sc">==</span> <span class="cn">TRUE</span>)))</span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">length</span>(nm)<span class="sc">&gt;</span> <span class="dv">0</span>){ db <span class="ot">&lt;-</span> db[<span class="sc">-</span>nm, ]}</span>
<span id="cb13-17"><a href="#cb13-17" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(nm)</span>
<span id="cb13-18"><a href="#cb13-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-19"><a href="#cb13-19" aria-hidden="true" tabindex="-1"></a><span class="co"># Separate into interactions with a single and multiple Ensembl Gene IDs</span></span>
<span id="cb13-20"><a href="#cb13-20" aria-hidden="true" tabindex="-1"></a>db_m2m <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">c</span>(<span class="fu">which</span>(<span class="fu">str_detect</span>(<span class="at">string =</span> db<span class="sc">$</span>Ensembl_Gene_Interactor_A, <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">|"</span>) <span class="sc">==</span> <span class="cn">TRUE</span>),</span>
<span id="cb13-21"><a href="#cb13-21" aria-hidden="true" tabindex="-1"></a>                   <span class="fu">which</span>(<span class="fu">str_detect</span>(<span class="at">string =</span> db<span class="sc">$</span>Ensembl_Gene_Interactor_B, <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">|"</span>) <span class="sc">==</span> <span class="cn">TRUE</span>)))</span>
<span id="cb13-22"><a href="#cb13-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-23"><a href="#cb13-23" aria-hidden="true" tabindex="-1"></a>db_o2o <span class="ot">&lt;-</span> <span class="fu">setdiff</span>(<span class="at">x =</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(db), <span class="at">y =</span> db_m2m)</span>
<span id="cb13-24"><a href="#cb13-24" aria-hidden="true" tabindex="-1"></a>db_m2m <span class="ot">&lt;-</span> db[db_m2m, ]</span>
<span id="cb13-25"><a href="#cb13-25" aria-hidden="true" tabindex="-1"></a>db_o2o <span class="ot">&lt;-</span> db[db_o2o, ]</span>
<span id="cb13-26"><a href="#cb13-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-27"><a href="#cb13-27" aria-hidden="true" tabindex="-1"></a><span class="co"># Expand interactions with multiple ensembl gene IDs</span></span>
<span id="cb13-28"><a href="#cb13-28" aria-hidden="true" tabindex="-1"></a>output <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="cn">NA</span>, <span class="at">nrow =</span> <span class="dv">1</span>, <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">dimnames =</span> <span class="fu">list</span>(<span class="st">"ToRemove"</span>, <span class="fu">c</span>(<span class="st">"Ensembl_Gene_Interactor_A"</span>, <span class="st">"Ensembl_Gene_Interactor_B"</span>)))</span>
<span id="cb13-29"><a href="#cb13-29" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(db_m2m))){</span>
<span id="cb13-30"><a href="#cb13-30" aria-hidden="true" tabindex="-1"></a>  in_a <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">str_split</span>(<span class="at">string =</span> db_m2m<span class="sc">$</span>Ensembl_Gene_Interactor_A[i], <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">|</span><span class="sc">\\</span><span class="st">|"</span>))</span>
<span id="cb13-31"><a href="#cb13-31" aria-hidden="true" tabindex="-1"></a>  in_b <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">str_split</span>(<span class="at">string =</span> db_m2m<span class="sc">$</span>Ensembl_Gene_Interactor_B[i], <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">|</span><span class="sc">\\</span><span class="st">|"</span>))</span>
<span id="cb13-32"><a href="#cb13-32" aria-hidden="true" tabindex="-1"></a>  ab <span class="ot">&lt;-</span> <span class="fu">expand.grid</span>(<span class="st">"Ensembl_Gene_Interactor_A"</span> <span class="ot">=</span> in_a, <span class="st">"Ensembl_Gene_Interactor_B"</span> <span class="ot">=</span> in_b)</span>
<span id="cb13-33"><a href="#cb13-33" aria-hidden="true" tabindex="-1"></a>  output <span class="ot">&lt;-</span> <span class="fu">rbind</span>(output, ab)</span>
<span id="cb13-34"><a href="#cb13-34" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb13-35"><a href="#cb13-35" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(in_a, in_b, ab, i)</span>
<span id="cb13-36"><a href="#cb13-36" aria-hidden="true" tabindex="-1"></a>output <span class="ot">&lt;-</span> output[<span class="sc">-</span><span class="dv">1</span>, ]</span>
<span id="cb13-37"><a href="#cb13-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-38"><a href="#cb13-38" aria-hidden="true" tabindex="-1"></a><span class="co"># Combine interactions that map a single and multiple Ensembl Gene IDs </span></span>
<span id="cb13-39"><a href="#cb13-39" aria-hidden="true" tabindex="-1"></a>db_ens <span class="ot">&lt;-</span> <span class="fu">rbind</span>(db_o2o[, <span class="fu">c</span>(<span class="st">"Ensembl_Gene_Interactor_A"</span>, <span class="st">"Ensembl_Gene_Interactor_B"</span>)], </span>
<span id="cb13-40"><a href="#cb13-40" aria-hidden="true" tabindex="-1"></a>                output)</span>
<span id="cb13-41"><a href="#cb13-41" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(db, db_m2m, db_o2o, output)</span>
<span id="cb13-42"><a href="#cb13-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-43"><a href="#cb13-43" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove self interactions</span></span>
<span id="cb13-44"><a href="#cb13-44" aria-hidden="true" tabindex="-1"></a>mark_self <span class="ot">&lt;-</span> <span class="fu">apply</span>(db_ens, <span class="dv">1</span>, <span class="at">FUN =</span> <span class="cf">function</span>(x) <span class="fu">length</span>(<span class="fu">unique</span>(x)))</span>
<span id="cb13-45"><a href="#cb13-45" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">length</span>(<span class="fu">which</span>(mark_self <span class="sc">==</span> <span class="dv">1</span>))<span class="sc">&gt;</span> <span class="dv">0</span>){db_ens <span class="ot">&lt;-</span> db_ens[<span class="sc">-</span><span class="fu">which</span>(mark_self <span class="sc">==</span> <span class="dv">1</span>), ]}</span>
<span id="cb13-46"><a href="#cb13-46" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(mark_self)</span>
<span id="cb13-47"><a href="#cb13-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-48"><a href="#cb13-48" aria-hidden="true" tabindex="-1"></a><span class="co"># Re-order interactions to remove duplicates</span></span>
<span id="cb13-49"><a href="#cb13-49" aria-hidden="true" tabindex="-1"></a>db_ens_order <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">apply</span>(db_ens, <span class="dv">1</span>, <span class="at">FUN =</span> <span class="cf">function</span>(x) x[<span class="fu">order</span>(x)]))</span>
<span id="cb13-50"><a href="#cb13-50" aria-hidden="true" tabindex="-1"></a>db_ens <span class="ot">&lt;-</span> <span class="fu">unique</span>(db_ens_order) </span>
<span id="cb13-51"><a href="#cb13-51" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(db_ens_order)</span>
<span id="cb13-52"><a href="#cb13-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-53"><a href="#cb13-53" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert to data frame</span></span>
<span id="cb13-54"><a href="#cb13-54" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(db_ens) <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(db_ens)</span>
<span id="cb13-55"><a href="#cb13-55" aria-hidden="true" tabindex="-1"></a>db_ens <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(db_ens)</span>
<span id="cb13-56"><a href="#cb13-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-57"><a href="#cb13-57" aria-hidden="true" tabindex="-1"></a><span class="co"># Save as tab delimited file </span></span>
<span id="cb13-58"><a href="#cb13-58" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(<span class="at">x =</span> db_ens, </span>
<span id="cb13-59"><a href="#cb13-59" aria-hidden="true" tabindex="-1"></a>            <span class="at">file =</span> <span class="st">"./02_Datasets/BIOPLEX/BioPlex_293T_Network_10K_Dec_2019_EnsemblGeneID.txt"</span>, </span>
<span id="cb13-60"><a href="#cb13-60" aria-hidden="true" tabindex="-1"></a>            <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>, <span class="at">col.names =</span> <span class="cn">FALSE</span>, <span class="at">quote =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="hct116" class="level3">
<h3 class="anchored" data-anchor-id="hct116">HCT116</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Bioplex</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>db <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="at">file =</span> <span class="st">"./02_Datasets/BIOPLEX/BioPlex_HCT116_Network_5.5K_Dec_2019.tsv"</span>, <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">header =</span> <span class="cn">TRUE</span>)</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove isoform info to mapp Uniprot ID</span></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>db<span class="sc">$</span>UniprotA <span class="ot">&lt;-</span> <span class="fu">sapply</span>(<span class="fu">str_split</span>(<span class="at">string =</span> db<span class="sc">$</span>UniprotA, <span class="at">pattern =</span> <span class="st">"-"</span>), <span class="st">"["</span>, <span class="dv">1</span>)</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>db<span class="sc">$</span>UniprotB <span class="ot">&lt;-</span> <span class="fu">sapply</span>(<span class="fu">str_split</span>(<span class="at">string =</span> db<span class="sc">$</span>UniprotB, <span class="at">pattern =</span> <span class="st">"-"</span>), <span class="st">"["</span>, <span class="dv">1</span>)</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Mapping through uniprot</span></span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>db <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(db, </span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>                 <span class="at">Ensembl_Gene_Interactor_A =</span> uniprot_to_ensembl<span class="sc">$</span>Ensembl_GeneID[<span class="fu">match</span>(<span class="at">x =</span> db<span class="sc">$</span>UniprotA, <span class="at">table =</span> uniprot_to_ensembl<span class="sc">$</span>Uniprot_ID)],</span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a>                 <span class="at">Ensembl_Gene_Interactor_B =</span> uniprot_to_ensembl<span class="sc">$</span>Ensembl_GeneID[<span class="fu">match</span>(<span class="at">x =</span> db<span class="sc">$</span>UniprotB, <span class="at">table =</span> uniprot_to_ensembl<span class="sc">$</span>Uniprot_ID)])</span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove interactions where either interactor cannot be mapped to Ensembl Gene IDs</span></span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true" tabindex="-1"></a>nm <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">c</span>(<span class="fu">which</span>(<span class="fu">is.na</span>(db<span class="sc">$</span>Ensembl_Gene_Interactor_A) <span class="sc">==</span> <span class="cn">TRUE</span>),</span>
<span id="cb14-15"><a href="#cb14-15" aria-hidden="true" tabindex="-1"></a>               <span class="fu">which</span>(<span class="fu">is.na</span>(db<span class="sc">$</span>Ensembl_Gene_Interactor_B) <span class="sc">==</span> <span class="cn">TRUE</span>)))</span>
<span id="cb14-16"><a href="#cb14-16" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">length</span>(nm)<span class="sc">&gt;</span> <span class="dv">0</span>){ db <span class="ot">&lt;-</span> db[<span class="sc">-</span>nm, ]}</span>
<span id="cb14-17"><a href="#cb14-17" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(nm)</span>
<span id="cb14-18"><a href="#cb14-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-19"><a href="#cb14-19" aria-hidden="true" tabindex="-1"></a><span class="co"># Separate into interactions with a single and multiple Ensembl Gene IDs</span></span>
<span id="cb14-20"><a href="#cb14-20" aria-hidden="true" tabindex="-1"></a>db_m2m <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">c</span>(<span class="fu">which</span>(<span class="fu">str_detect</span>(<span class="at">string =</span> db<span class="sc">$</span>Ensembl_Gene_Interactor_A, <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">|"</span>) <span class="sc">==</span> <span class="cn">TRUE</span>),</span>
<span id="cb14-21"><a href="#cb14-21" aria-hidden="true" tabindex="-1"></a>                   <span class="fu">which</span>(<span class="fu">str_detect</span>(<span class="at">string =</span> db<span class="sc">$</span>Ensembl_Gene_Interactor_B, <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">|"</span>) <span class="sc">==</span> <span class="cn">TRUE</span>)))</span>
<span id="cb14-22"><a href="#cb14-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-23"><a href="#cb14-23" aria-hidden="true" tabindex="-1"></a>db_o2o <span class="ot">&lt;-</span> <span class="fu">setdiff</span>(<span class="at">x =</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(db), <span class="at">y =</span> db_m2m)</span>
<span id="cb14-24"><a href="#cb14-24" aria-hidden="true" tabindex="-1"></a>db_m2m <span class="ot">&lt;-</span> db[db_m2m, ]</span>
<span id="cb14-25"><a href="#cb14-25" aria-hidden="true" tabindex="-1"></a>db_o2o <span class="ot">&lt;-</span> db[db_o2o, ]</span>
<span id="cb14-26"><a href="#cb14-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-27"><a href="#cb14-27" aria-hidden="true" tabindex="-1"></a><span class="co"># Expand interactions with multiple ensembl gene IDs</span></span>
<span id="cb14-28"><a href="#cb14-28" aria-hidden="true" tabindex="-1"></a>output <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="cn">NA</span>, <span class="at">nrow =</span> <span class="dv">1</span>, <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">dimnames =</span> <span class="fu">list</span>(<span class="st">"ToRemove"</span>, <span class="fu">c</span>(<span class="st">"Ensembl_Gene_Interactor_A"</span>, <span class="st">"Ensembl_Gene_Interactor_B"</span>)))</span>
<span id="cb14-29"><a href="#cb14-29" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(db_m2m))){</span>
<span id="cb14-30"><a href="#cb14-30" aria-hidden="true" tabindex="-1"></a>  in_a <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">str_split</span>(<span class="at">string =</span> db_m2m<span class="sc">$</span>Ensembl_Gene_Interactor_A[i], <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">|</span><span class="sc">\\</span><span class="st">|"</span>))</span>
<span id="cb14-31"><a href="#cb14-31" aria-hidden="true" tabindex="-1"></a>  in_b <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">str_split</span>(<span class="at">string =</span> db_m2m<span class="sc">$</span>Ensembl_Gene_Interactor_B[i], <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">|</span><span class="sc">\\</span><span class="st">|"</span>))</span>
<span id="cb14-32"><a href="#cb14-32" aria-hidden="true" tabindex="-1"></a>  ab <span class="ot">&lt;-</span> <span class="fu">expand.grid</span>(<span class="st">"Ensembl_Gene_Interactor_A"</span> <span class="ot">=</span> in_a, <span class="st">"Ensembl_Gene_Interactor_B"</span> <span class="ot">=</span> in_b)</span>
<span id="cb14-33"><a href="#cb14-33" aria-hidden="true" tabindex="-1"></a>  output <span class="ot">&lt;-</span> <span class="fu">rbind</span>(output, ab)</span>
<span id="cb14-34"><a href="#cb14-34" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb14-35"><a href="#cb14-35" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(in_a, in_b, ab, i)</span>
<span id="cb14-36"><a href="#cb14-36" aria-hidden="true" tabindex="-1"></a>output <span class="ot">&lt;-</span> output[<span class="sc">-</span><span class="dv">1</span>, ]</span>
<span id="cb14-37"><a href="#cb14-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-38"><a href="#cb14-38" aria-hidden="true" tabindex="-1"></a><span class="co"># Combine interactions that map a single and multiple Ensembl Gene IDs </span></span>
<span id="cb14-39"><a href="#cb14-39" aria-hidden="true" tabindex="-1"></a>db_ens <span class="ot">&lt;-</span> <span class="fu">rbind</span>(db_o2o[, <span class="fu">c</span>(<span class="st">"Ensembl_Gene_Interactor_A"</span>, <span class="st">"Ensembl_Gene_Interactor_B"</span>)], </span>
<span id="cb14-40"><a href="#cb14-40" aria-hidden="true" tabindex="-1"></a>                output)</span>
<span id="cb14-41"><a href="#cb14-41" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(db, db_m2m, db_o2o, output)</span>
<span id="cb14-42"><a href="#cb14-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-43"><a href="#cb14-43" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove self interactions</span></span>
<span id="cb14-44"><a href="#cb14-44" aria-hidden="true" tabindex="-1"></a>mark_self <span class="ot">&lt;-</span> <span class="fu">apply</span>(db_ens, <span class="dv">1</span>, <span class="at">FUN =</span> <span class="cf">function</span>(x) <span class="fu">length</span>(<span class="fu">unique</span>(x)))</span>
<span id="cb14-45"><a href="#cb14-45" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">length</span>(<span class="fu">which</span>(mark_self <span class="sc">==</span> <span class="dv">1</span>))<span class="sc">&gt;</span> <span class="dv">0</span>){db_ens <span class="ot">&lt;-</span> db_ens[<span class="sc">-</span><span class="fu">which</span>(mark_self <span class="sc">==</span> <span class="dv">1</span>), ]}</span>
<span id="cb14-46"><a href="#cb14-46" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(mark_self)</span>
<span id="cb14-47"><a href="#cb14-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-48"><a href="#cb14-48" aria-hidden="true" tabindex="-1"></a><span class="co"># Re-order interactions to remove duplicates</span></span>
<span id="cb14-49"><a href="#cb14-49" aria-hidden="true" tabindex="-1"></a>db_ens_order <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">apply</span>(db_ens, <span class="dv">1</span>, <span class="at">FUN =</span> <span class="cf">function</span>(x) x[<span class="fu">order</span>(x)]))</span>
<span id="cb14-50"><a href="#cb14-50" aria-hidden="true" tabindex="-1"></a>db_ens <span class="ot">&lt;-</span> <span class="fu">unique</span>(db_ens_order) </span>
<span id="cb14-51"><a href="#cb14-51" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(db_ens_order)</span>
<span id="cb14-52"><a href="#cb14-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-53"><a href="#cb14-53" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert to data frame</span></span>
<span id="cb14-54"><a href="#cb14-54" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(db_ens) <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(db_ens)</span>
<span id="cb14-55"><a href="#cb14-55" aria-hidden="true" tabindex="-1"></a>db_ens <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(db_ens)</span>
<span id="cb14-56"><a href="#cb14-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-57"><a href="#cb14-57" aria-hidden="true" tabindex="-1"></a><span class="co"># Save as tab delimited file </span></span>
<span id="cb14-58"><a href="#cb14-58" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(<span class="at">x =</span> db_ens, </span>
<span id="cb14-59"><a href="#cb14-59" aria-hidden="true" tabindex="-1"></a>            <span class="at">file =</span> <span class="st">"./02_Datasets/BIOPLEX/BioPlex_HCT116_Network_5.5K_Dec_2019_EnsemblGeneID.txt"</span>, </span>
<span id="cb14-60"><a href="#cb14-60" aria-hidden="true" tabindex="-1"></a>            <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>, <span class="at">col.names =</span> <span class="cn">FALSE</span>, <span class="at">quote =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="corum" class="level2">
<h2 class="anchored" data-anchor-id="corum">CORUM</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="st">"./02_Datasets/CORUM/"</span>)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Database is now (Feb 2024) online !</span></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a><span class="co"># All complexes</span></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(<span class="st">"https://mips.helmholtz-muenchen.de/corum/download/releases/current/allComplexes.txt.zip"</span>, </span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>              <span class="at">destfile =</span> <span class="st">"./02_Datasets/CORUM/allComplex.zip"</span>)</span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Human complexes</span></span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(<span class="st">"https://mips.helmholtz-muenchen.de/corum/download/releases/current/humanComplexes.txt.zip"</span>, </span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a>              <span class="at">destfile =</span> <span class="st">"./02_Datasets/CORUM/humanComplexes.txt.zip"</span>)</span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Partial complexes</span></span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(<span class="st">"https://mips.helmholtz-muenchen.de/corum/download/releases/current/partialComplexes.txt.zip"</span>, </span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a>              <span class="at">destfile =</span> <span class="st">"./02_Datasets/CORUM/partialComplexes.txt.zip"</span>)</span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Unzip files</span></span>
<span id="cb15-18"><a href="#cb15-18" aria-hidden="true" tabindex="-1"></a><span class="fu">gunzip</span>(<span class="at">zipfile =</span> <span class="st">"./02_Datasets/CORUM/allComplexes.txt.zip"</span>, <span class="at">remove =</span> <span class="cn">TRUE</span>)</span>
<span id="cb15-19"><a href="#cb15-19" aria-hidden="true" tabindex="-1"></a><span class="fu">gunzip</span>(<span class="at">zipfile =</span> <span class="st">"./02_Datasets/CORUM/humanComplexes.txt.zip"</span>, <span class="at">remove =</span> <span class="cn">TRUE</span>)</span>
<span id="cb15-20"><a href="#cb15-20" aria-hidden="true" tabindex="-1"></a><span class="fu">gunzip</span>(<span class="at">zipfile =</span> <span class="st">"./02_Datasets/CORUM/partialComplexes.txt.zip"</span>, <span class="at">remove =</span> <span class="cn">TRUE</span>)</span>
<span id="cb15-21"><a href="#cb15-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-22"><a href="#cb15-22" aria-hidden="true" tabindex="-1"></a><span class="fu">Sys.setenv</span>(<span class="at">CORUM_ALL =</span> <span class="st">"./02_Datasets/CORUM/allComplexes.txt"</span>)</span>
<span id="cb15-23"><a href="#cb15-23" aria-hidden="true" tabindex="-1"></a><span class="fu">Sys.setenv</span>(<span class="at">CORUM_ALL_TMP =</span> <span class="st">"./02_Datasets/CORUM/allComplexes_tmp.txt"</span>)</span>
<span id="cb15-24"><a href="#cb15-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-25"><a href="#cb15-25" aria-hidden="true" tabindex="-1"></a><span class="fu">Sys.setenv</span>(<span class="at">CORUM_HUM =</span> <span class="st">"./02_Datasets/CORUM/humanComplexes.txt"</span>)</span>
<span id="cb15-26"><a href="#cb15-26" aria-hidden="true" tabindex="-1"></a><span class="fu">Sys.setenv</span>(<span class="at">CORUM_HUM_TMP =</span> <span class="st">"./02_Datasets/CORUM/humanComplexes_tmp.txt"</span>)</span>
<span id="cb15-27"><a href="#cb15-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-28"><a href="#cb15-28" aria-hidden="true" tabindex="-1"></a><span class="fu">Sys.setenv</span>(<span class="at">CORUM_PAR =</span> <span class="st">"./02_Datasets/CORUM/partialComplexes.txt"</span>)</span>
<span id="cb15-29"><a href="#cb15-29" aria-hidden="true" tabindex="-1"></a><span class="fu">Sys.setenv</span>(<span class="at">CORUM_PAR_TMP =</span> <span class="st">"./02_Datasets/CORUM/partialComplexes_tmp.txt"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Parse file to remove ' from file</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a><span class="fu">sed</span> <span class="st">"s/'//g"</span> <span class="va">$CORUM_ALL</span> <span class="op">&gt;</span> <span class="va">$CORUM_ALL_TMP</span></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a><span class="fu">sed</span> <span class="st">"s/'//g"</span> <span class="va">$CORUM_HUM</span> <span class="op">&gt;</span> <span class="va">$CORUM_HUM_TMP</span></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a><span class="fu">sed</span> <span class="st">"s/'//g"</span> <span class="va">$CORUM_PAR</span> <span class="op">&gt;</span> <span class="va">$CORUM_PAR_TMP</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove temporal file and rename</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="fu">unlink</span>(<span class="at">x =</span> <span class="st">"./02_Datasets/CORUM/allComplexes.txt"</span>)</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="fu">unlink</span>(<span class="at">x =</span> <span class="st">"./02_Datasets/CORUM/humanComplexes.txt"</span>)</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a><span class="fu">unlink</span>(<span class="at">x =</span> <span class="st">"./02_Datasets/CORUM/partialComplexes.txt"</span>)</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a><span class="fu">file.rename</span>(<span class="at">from =</span> <span class="st">"./02_Datasets/CORUM/allComplexes_tmp.txt"</span>, <span class="at">to =</span> <span class="st">"./02_Datasets/CORUM/allComplexes.txt"</span>)</span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a><span class="fu">file.rename</span>(<span class="at">from =</span> <span class="st">"./02_Datasets/CORUM/humanComplexes_tmp.txt"</span>, <span class="at">to =</span> <span class="st">"./02_Datasets/CORUM/humanComplexes.txt"</span>)</span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a><span class="fu">file.rename</span>(<span class="at">from =</span> <span class="st">"./02_Datasets/CORUM/partialComplexes_tmp.txt"</span>, <span class="at">to =</span> <span class="st">"./02_Datasets/CORUM/partialComplexes.txt"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="all-complexes" class="level3">
<h3 class="anchored" data-anchor-id="all-complexes">All complexes</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># CORUM</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>db <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="at">file =</span> <span class="st">"./02_Datasets/CORUM/allComplexes.txt"</span>, <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">header =</span> <span class="cn">TRUE</span>)</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Rename column names </span></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(db) <span class="ot">&lt;-</span> <span class="fu">str_replace_all</span>(<span class="at">string =</span> <span class="fu">colnames</span>(db), <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">."</span>, <span class="at">replacement =</span> <span class="st">"_"</span>)</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Reduce to human interactions</span></span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>db <span class="ot">&lt;-</span> db[<span class="fu">which</span>(db<span class="sc">$</span>Organism <span class="sc">==</span> <span class="st">"Human"</span>), ]</span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a><span class="co"># From complexes to pairs of interactions</span></span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a>output <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="cn">NA</span>, <span class="at">nrow =</span> <span class="dv">1</span>, <span class="at">ncol =</span> <span class="dv">2</span>, </span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a>                 <span class="at">dimnames =</span> <span class="fu">list</span>(<span class="st">"RemoveMe"</span>, <span class="fu">c</span>(<span class="st">"Interactor_A"</span>, <span class="st">"Interactor_B"</span>)))</span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(db)){</span>
<span id="cb18-15"><a href="#cb18-15" aria-hidden="true" tabindex="-1"></a>  ci <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">str_split</span>(<span class="at">string =</span> db<span class="sc">$</span>subunits_UniProt_IDs_[i], <span class="at">pattern =</span> <span class="st">";"</span>))</span>
<span id="cb18-16"><a href="#cb18-16" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">length</span>(ci) <span class="sc">&gt;</span> <span class="dv">0</span>){</span>
<span id="cb18-17"><a href="#cb18-17" aria-hidden="true" tabindex="-1"></a>  output <span class="ot">&lt;-</span> <span class="fu">rbind</span>(output, <span class="fu">expand.grid</span>(<span class="st">"Interactor_A"</span> <span class="ot">=</span> ci, <span class="st">"Interactor_B"</span> <span class="ot">=</span> ci))  </span>
<span id="cb18-18"><a href="#cb18-18" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb18-19"><a href="#cb18-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rm</span>(ci)</span>
<span id="cb18-20"><a href="#cb18-20" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb18-21"><a href="#cb18-21" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(i)</span>
<span id="cb18-22"><a href="#cb18-22" aria-hidden="true" tabindex="-1"></a>output <span class="ot">&lt;-</span> output[<span class="sc">-</span><span class="dv">1</span>, ]</span>
<span id="cb18-23"><a href="#cb18-23" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(db)</span>
<span id="cb18-24"><a href="#cb18-24" aria-hidden="true" tabindex="-1"></a>db <span class="ot">&lt;-</span> output</span>
<span id="cb18-25"><a href="#cb18-25" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(output)</span>
<span id="cb18-26"><a href="#cb18-26" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(db) <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(db)</span>
<span id="cb18-27"><a href="#cb18-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-28"><a href="#cb18-28" aria-hidden="true" tabindex="-1"></a><span class="co"># Add ensembl annotations</span></span>
<span id="cb18-29"><a href="#cb18-29" aria-hidden="true" tabindex="-1"></a>db <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(db, </span>
<span id="cb18-30"><a href="#cb18-30" aria-hidden="true" tabindex="-1"></a>                 <span class="at">Ensembl_Gene_Interactor_A =</span> uniprot_to_ensembl<span class="sc">$</span>Ensembl_GeneID[<span class="fu">match</span>(<span class="at">x =</span> db<span class="sc">$</span>Interactor_A, <span class="at">table =</span> uniprot_to_ensembl<span class="sc">$</span>Uniprot_ID)],</span>
<span id="cb18-31"><a href="#cb18-31" aria-hidden="true" tabindex="-1"></a>                 <span class="at">Ensembl_Gene_Interactor_B =</span> uniprot_to_ensembl<span class="sc">$</span>Ensembl_GeneID[<span class="fu">match</span>(<span class="at">x =</span> db<span class="sc">$</span>Interactor_B, <span class="at">table =</span> uniprot_to_ensembl<span class="sc">$</span>Uniprot_ID)])</span>
<span id="cb18-32"><a href="#cb18-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-33"><a href="#cb18-33" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove interactions where either interactor cannot be mapped to Ensembl Gene IDs</span></span>
<span id="cb18-34"><a href="#cb18-34" aria-hidden="true" tabindex="-1"></a>nm <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">c</span>(<span class="fu">which</span>(<span class="fu">is.na</span>(db<span class="sc">$</span>Ensembl_Gene_Interactor_A) <span class="sc">==</span> <span class="cn">TRUE</span>),</span>
<span id="cb18-35"><a href="#cb18-35" aria-hidden="true" tabindex="-1"></a>               <span class="fu">which</span>(<span class="fu">is.na</span>(db<span class="sc">$</span>Ensembl_Gene_Interactor_B) <span class="sc">==</span> <span class="cn">TRUE</span>)))</span>
<span id="cb18-36"><a href="#cb18-36" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">length</span>(nm)<span class="sc">&gt;</span> <span class="dv">0</span>){ db <span class="ot">&lt;-</span> db[<span class="sc">-</span>nm, ]}</span>
<span id="cb18-37"><a href="#cb18-37" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(nm)</span>
<span id="cb18-38"><a href="#cb18-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-39"><a href="#cb18-39" aria-hidden="true" tabindex="-1"></a><span class="co"># Separate into interactions with a single and multiple Ensembl Gene IDs</span></span>
<span id="cb18-40"><a href="#cb18-40" aria-hidden="true" tabindex="-1"></a>db_m2m <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">c</span>(<span class="fu">which</span>(<span class="fu">str_detect</span>(<span class="at">string =</span> db<span class="sc">$</span>Ensembl_Gene_Interactor_A, <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">|"</span>) <span class="sc">==</span> <span class="cn">TRUE</span>),</span>
<span id="cb18-41"><a href="#cb18-41" aria-hidden="true" tabindex="-1"></a>                   <span class="fu">which</span>(<span class="fu">str_detect</span>(<span class="at">string =</span> db<span class="sc">$</span>Ensembl_Gene_Interactor_B, <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">|"</span>) <span class="sc">==</span> <span class="cn">TRUE</span>)))</span>
<span id="cb18-42"><a href="#cb18-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-43"><a href="#cb18-43" aria-hidden="true" tabindex="-1"></a>db_o2o <span class="ot">&lt;-</span> <span class="fu">setdiff</span>(<span class="at">x =</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(db), <span class="at">y =</span> db_m2m)</span>
<span id="cb18-44"><a href="#cb18-44" aria-hidden="true" tabindex="-1"></a>db_m2m <span class="ot">&lt;-</span> db[db_m2m, ]</span>
<span id="cb18-45"><a href="#cb18-45" aria-hidden="true" tabindex="-1"></a>db_o2o <span class="ot">&lt;-</span> db[db_o2o, ]</span>
<span id="cb18-46"><a href="#cb18-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-47"><a href="#cb18-47" aria-hidden="true" tabindex="-1"></a><span class="co"># Expand interactions with multiple ensembl gene IDs</span></span>
<span id="cb18-48"><a href="#cb18-48" aria-hidden="true" tabindex="-1"></a>output <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="cn">NA</span>, <span class="at">nrow =</span> <span class="dv">1</span>, <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">dimnames =</span> <span class="fu">list</span>(<span class="st">"ToRemove"</span>, <span class="fu">c</span>(<span class="st">"Ensembl_Gene_Interactor_A"</span>, <span class="st">"Ensembl_Gene_Interactor_B"</span>)))</span>
<span id="cb18-49"><a href="#cb18-49" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(db_m2m))){</span>
<span id="cb18-50"><a href="#cb18-50" aria-hidden="true" tabindex="-1"></a>  in_a <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">str_split</span>(<span class="at">string =</span> db_m2m<span class="sc">$</span>Ensembl_Gene_Interactor_A[i], <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">|</span><span class="sc">\\</span><span class="st">|"</span>))</span>
<span id="cb18-51"><a href="#cb18-51" aria-hidden="true" tabindex="-1"></a>  in_b <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">str_split</span>(<span class="at">string =</span> db_m2m<span class="sc">$</span>Ensembl_Gene_Interactor_B[i], <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">|</span><span class="sc">\\</span><span class="st">|"</span>))</span>
<span id="cb18-52"><a href="#cb18-52" aria-hidden="true" tabindex="-1"></a>  ab <span class="ot">&lt;-</span> <span class="fu">expand.grid</span>(<span class="st">"Ensembl_Gene_Interactor_A"</span> <span class="ot">=</span> in_a, <span class="st">"Ensembl_Gene_Interactor_B"</span> <span class="ot">=</span> in_b)</span>
<span id="cb18-53"><a href="#cb18-53" aria-hidden="true" tabindex="-1"></a>  output <span class="ot">&lt;-</span> <span class="fu">rbind</span>(output, ab)</span>
<span id="cb18-54"><a href="#cb18-54" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb18-55"><a href="#cb18-55" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(in_a, in_b, ab, i)</span>
<span id="cb18-56"><a href="#cb18-56" aria-hidden="true" tabindex="-1"></a>output <span class="ot">&lt;-</span> output[<span class="sc">-</span><span class="dv">1</span>, ]</span>
<span id="cb18-57"><a href="#cb18-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-58"><a href="#cb18-58" aria-hidden="true" tabindex="-1"></a><span class="co"># Combine interactions that map a single and multiple Ensembl Gene IDs </span></span>
<span id="cb18-59"><a href="#cb18-59" aria-hidden="true" tabindex="-1"></a>db_ens <span class="ot">&lt;-</span> <span class="fu">rbind</span>(db_o2o[, <span class="fu">c</span>(<span class="st">"Ensembl_Gene_Interactor_A"</span>, <span class="st">"Ensembl_Gene_Interactor_B"</span>)], </span>
<span id="cb18-60"><a href="#cb18-60" aria-hidden="true" tabindex="-1"></a>                output)</span>
<span id="cb18-61"><a href="#cb18-61" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(db, db_m2m, db_o2o, output)</span>
<span id="cb18-62"><a href="#cb18-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-63"><a href="#cb18-63" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove self interactions</span></span>
<span id="cb18-64"><a href="#cb18-64" aria-hidden="true" tabindex="-1"></a>mark_self <span class="ot">&lt;-</span> <span class="fu">apply</span>(db_ens, <span class="dv">1</span>, <span class="at">FUN =</span> <span class="cf">function</span>(x) <span class="fu">length</span>(<span class="fu">unique</span>(x)))</span>
<span id="cb18-65"><a href="#cb18-65" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">length</span>(<span class="fu">which</span>(mark_self <span class="sc">==</span> <span class="dv">1</span>))<span class="sc">&gt;</span> <span class="dv">0</span>){db_ens <span class="ot">&lt;-</span> db_ens[<span class="sc">-</span><span class="fu">which</span>(mark_self <span class="sc">==</span> <span class="dv">1</span>), ]}</span>
<span id="cb18-66"><a href="#cb18-66" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(mark_self)</span>
<span id="cb18-67"><a href="#cb18-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-68"><a href="#cb18-68" aria-hidden="true" tabindex="-1"></a><span class="co"># Re-order interactions to remove duplicates</span></span>
<span id="cb18-69"><a href="#cb18-69" aria-hidden="true" tabindex="-1"></a>db_ens_order <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">apply</span>(db_ens, <span class="dv">1</span>, <span class="at">FUN =</span> <span class="cf">function</span>(x) x[<span class="fu">order</span>(x)]))</span>
<span id="cb18-70"><a href="#cb18-70" aria-hidden="true" tabindex="-1"></a>db_ens <span class="ot">&lt;-</span> <span class="fu">unique</span>(db_ens_order) </span>
<span id="cb18-71"><a href="#cb18-71" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(db_ens_order)</span>
<span id="cb18-72"><a href="#cb18-72" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-73"><a href="#cb18-73" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert to data frame</span></span>
<span id="cb18-74"><a href="#cb18-74" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(db_ens) <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(db_ens)</span>
<span id="cb18-75"><a href="#cb18-75" aria-hidden="true" tabindex="-1"></a>db_ens <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(db_ens)</span>
<span id="cb18-76"><a href="#cb18-76" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-77"><a href="#cb18-77" aria-hidden="true" tabindex="-1"></a><span class="co"># Save as tab delimited file </span></span>
<span id="cb18-78"><a href="#cb18-78" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(<span class="at">x =</span> db_ens, </span>
<span id="cb18-79"><a href="#cb18-79" aria-hidden="true" tabindex="-1"></a>            <span class="at">file =</span> <span class="st">"./02_Datasets/CORUM/allComplexes_EnsemblGeneID.txt"</span>, </span>
<span id="cb18-80"><a href="#cb18-80" aria-hidden="true" tabindex="-1"></a>            <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>, <span class="at">col.names =</span> <span class="cn">FALSE</span>, <span class="at">quote =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="human-complexes" class="level3">
<h3 class="anchored" data-anchor-id="human-complexes">Human complexes</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># CORUM</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>db <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="at">file =</span> <span class="st">"./02_Datasets/CORUM/humanComplexes.txt"</span>, <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">header =</span> <span class="cn">TRUE</span>)</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Rename column names </span></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(db) <span class="ot">&lt;-</span> <span class="fu">str_replace_all</span>(<span class="at">string =</span> <span class="fu">colnames</span>(db), <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">."</span>, <span class="at">replacement =</span> <span class="st">"_"</span>)</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Reduce to human interactions</span></span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a>db <span class="ot">&lt;-</span> db[<span class="fu">which</span>(db<span class="sc">$</span>Organism <span class="sc">==</span> <span class="st">"Human"</span>), ]</span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a><span class="co"># From complexes to pairs of interactions</span></span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a>output <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="cn">NA</span>, <span class="at">nrow =</span> <span class="dv">1</span>, <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">dimnames =</span> <span class="fu">list</span>(<span class="st">"RemoveMe"</span>, <span class="fu">c</span>(<span class="st">"Interactor_A"</span>, <span class="st">"Interactor_B"</span>)))</span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(db)){</span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a>  ci <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">str_split</span>(<span class="at">string =</span> db<span class="sc">$</span>subunits_UniProt_IDs_[i], <span class="at">pattern =</span> <span class="st">";"</span>))</span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">length</span>(ci) <span class="sc">&gt;</span> <span class="dv">0</span>){</span>
<span id="cb19-15"><a href="#cb19-15" aria-hidden="true" tabindex="-1"></a>  output <span class="ot">&lt;-</span> <span class="fu">rbind</span>(output, <span class="fu">expand.grid</span>(<span class="st">"Interactor_A"</span> <span class="ot">=</span> ci, <span class="st">"Interactor_B"</span> <span class="ot">=</span> ci))  </span>
<span id="cb19-16"><a href="#cb19-16" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb19-17"><a href="#cb19-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rm</span>(ci)</span>
<span id="cb19-18"><a href="#cb19-18" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb19-19"><a href="#cb19-19" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(i)</span>
<span id="cb19-20"><a href="#cb19-20" aria-hidden="true" tabindex="-1"></a>output <span class="ot">&lt;-</span> output[<span class="sc">-</span><span class="dv">1</span>, ]</span>
<span id="cb19-21"><a href="#cb19-21" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(db)</span>
<span id="cb19-22"><a href="#cb19-22" aria-hidden="true" tabindex="-1"></a>db <span class="ot">&lt;-</span> output</span>
<span id="cb19-23"><a href="#cb19-23" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(output)</span>
<span id="cb19-24"><a href="#cb19-24" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(db) <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(db)</span>
<span id="cb19-25"><a href="#cb19-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-26"><a href="#cb19-26" aria-hidden="true" tabindex="-1"></a><span class="co"># Add ensembl annotations</span></span>
<span id="cb19-27"><a href="#cb19-27" aria-hidden="true" tabindex="-1"></a>db <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(db, </span>
<span id="cb19-28"><a href="#cb19-28" aria-hidden="true" tabindex="-1"></a>                 <span class="at">Ensembl_Gene_Interactor_A =</span> uniprot_to_ensembl<span class="sc">$</span>Ensembl_GeneID[<span class="fu">match</span>(<span class="at">x =</span> db<span class="sc">$</span>Interactor_A, <span class="at">table =</span> uniprot_to_ensembl<span class="sc">$</span>Uniprot_ID)],</span>
<span id="cb19-29"><a href="#cb19-29" aria-hidden="true" tabindex="-1"></a>                 <span class="at">Ensembl_Gene_Interactor_B =</span> uniprot_to_ensembl<span class="sc">$</span>Ensembl_GeneID[<span class="fu">match</span>(<span class="at">x =</span> db<span class="sc">$</span>Interactor_B, <span class="at">table =</span> uniprot_to_ensembl<span class="sc">$</span>Uniprot_ID)])</span>
<span id="cb19-30"><a href="#cb19-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-31"><a href="#cb19-31" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove interactions where either interactor cannot be mapped to Ensembl Gene IDs</span></span>
<span id="cb19-32"><a href="#cb19-32" aria-hidden="true" tabindex="-1"></a>nm <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">c</span>(<span class="fu">which</span>(<span class="fu">is.na</span>(db<span class="sc">$</span>Ensembl_Gene_Interactor_A) <span class="sc">==</span> <span class="cn">TRUE</span>),</span>
<span id="cb19-33"><a href="#cb19-33" aria-hidden="true" tabindex="-1"></a>               <span class="fu">which</span>(<span class="fu">is.na</span>(db<span class="sc">$</span>Ensembl_Gene_Interactor_B) <span class="sc">==</span> <span class="cn">TRUE</span>)))</span>
<span id="cb19-34"><a href="#cb19-34" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">length</span>(nm)<span class="sc">&gt;</span> <span class="dv">0</span>){ db <span class="ot">&lt;-</span> db[<span class="sc">-</span>nm, ]}</span>
<span id="cb19-35"><a href="#cb19-35" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(nm)</span>
<span id="cb19-36"><a href="#cb19-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-37"><a href="#cb19-37" aria-hidden="true" tabindex="-1"></a><span class="co"># Separate into interactions with a single and multiple Ensembl Gene IDs</span></span>
<span id="cb19-38"><a href="#cb19-38" aria-hidden="true" tabindex="-1"></a>db_m2m <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">c</span>(<span class="fu">which</span>(<span class="fu">str_detect</span>(<span class="at">string =</span> db<span class="sc">$</span>Ensembl_Gene_Interactor_A, <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">|"</span>) <span class="sc">==</span> <span class="cn">TRUE</span>),</span>
<span id="cb19-39"><a href="#cb19-39" aria-hidden="true" tabindex="-1"></a>                   <span class="fu">which</span>(<span class="fu">str_detect</span>(<span class="at">string =</span> db<span class="sc">$</span>Ensembl_Gene_Interactor_B, <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">|"</span>) <span class="sc">==</span> <span class="cn">TRUE</span>)))</span>
<span id="cb19-40"><a href="#cb19-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-41"><a href="#cb19-41" aria-hidden="true" tabindex="-1"></a>db_o2o <span class="ot">&lt;-</span> <span class="fu">setdiff</span>(<span class="at">x =</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(db), <span class="at">y =</span> db_m2m)</span>
<span id="cb19-42"><a href="#cb19-42" aria-hidden="true" tabindex="-1"></a>db_m2m <span class="ot">&lt;-</span> db[db_m2m, ]</span>
<span id="cb19-43"><a href="#cb19-43" aria-hidden="true" tabindex="-1"></a>db_o2o <span class="ot">&lt;-</span> db[db_o2o, ]</span>
<span id="cb19-44"><a href="#cb19-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-45"><a href="#cb19-45" aria-hidden="true" tabindex="-1"></a><span class="co"># Expand interactions with multiple ensembl gene IDs</span></span>
<span id="cb19-46"><a href="#cb19-46" aria-hidden="true" tabindex="-1"></a>output <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="cn">NA</span>, <span class="at">nrow =</span> <span class="dv">1</span>, <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">dimnames =</span> <span class="fu">list</span>(<span class="st">"ToRemove"</span>, <span class="fu">c</span>(<span class="st">"Ensembl_Gene_Interactor_A"</span>, <span class="st">"Ensembl_Gene_Interactor_B"</span>)))</span>
<span id="cb19-47"><a href="#cb19-47" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(db_m2m))){</span>
<span id="cb19-48"><a href="#cb19-48" aria-hidden="true" tabindex="-1"></a>  in_a <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">str_split</span>(<span class="at">string =</span> db_m2m<span class="sc">$</span>Ensembl_Gene_Interactor_A[i], <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">|</span><span class="sc">\\</span><span class="st">|"</span>))</span>
<span id="cb19-49"><a href="#cb19-49" aria-hidden="true" tabindex="-1"></a>  in_b <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">str_split</span>(<span class="at">string =</span> db_m2m<span class="sc">$</span>Ensembl_Gene_Interactor_B[i], <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">|</span><span class="sc">\\</span><span class="st">|"</span>))</span>
<span id="cb19-50"><a href="#cb19-50" aria-hidden="true" tabindex="-1"></a>  ab <span class="ot">&lt;-</span> <span class="fu">expand.grid</span>(<span class="st">"Ensembl_Gene_Interactor_A"</span> <span class="ot">=</span> in_a, <span class="st">"Ensembl_Gene_Interactor_B"</span> <span class="ot">=</span> in_b)</span>
<span id="cb19-51"><a href="#cb19-51" aria-hidden="true" tabindex="-1"></a>  output <span class="ot">&lt;-</span> <span class="fu">rbind</span>(output, ab)</span>
<span id="cb19-52"><a href="#cb19-52" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb19-53"><a href="#cb19-53" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(in_a, in_b, ab, i)</span>
<span id="cb19-54"><a href="#cb19-54" aria-hidden="true" tabindex="-1"></a>output <span class="ot">&lt;-</span> output[<span class="sc">-</span><span class="dv">1</span>, ]</span>
<span id="cb19-55"><a href="#cb19-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-56"><a href="#cb19-56" aria-hidden="true" tabindex="-1"></a><span class="co"># Combine interactions that map a single and multiple Ensembl Gene IDs </span></span>
<span id="cb19-57"><a href="#cb19-57" aria-hidden="true" tabindex="-1"></a>db_ens <span class="ot">&lt;-</span> <span class="fu">rbind</span>(db_o2o[, <span class="fu">c</span>(<span class="st">"Ensembl_Gene_Interactor_A"</span>, <span class="st">"Ensembl_Gene_Interactor_B"</span>)], </span>
<span id="cb19-58"><a href="#cb19-58" aria-hidden="true" tabindex="-1"></a>                output)</span>
<span id="cb19-59"><a href="#cb19-59" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(db, db_m2m, db_o2o, output)</span>
<span id="cb19-60"><a href="#cb19-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-61"><a href="#cb19-61" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove self interactions</span></span>
<span id="cb19-62"><a href="#cb19-62" aria-hidden="true" tabindex="-1"></a>mark_self <span class="ot">&lt;-</span> <span class="fu">apply</span>(db_ens, <span class="dv">1</span>, <span class="at">FUN =</span> <span class="cf">function</span>(x) <span class="fu">length</span>(<span class="fu">unique</span>(x)))</span>
<span id="cb19-63"><a href="#cb19-63" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">length</span>(<span class="fu">which</span>(mark_self <span class="sc">==</span> <span class="dv">1</span>))<span class="sc">&gt;</span> <span class="dv">0</span>){db_ens <span class="ot">&lt;-</span> db_ens[<span class="sc">-</span><span class="fu">which</span>(mark_self <span class="sc">==</span> <span class="dv">1</span>), ]}</span>
<span id="cb19-64"><a href="#cb19-64" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(mark_self)</span>
<span id="cb19-65"><a href="#cb19-65" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-66"><a href="#cb19-66" aria-hidden="true" tabindex="-1"></a><span class="co"># Re-order interactions to remove duplicates</span></span>
<span id="cb19-67"><a href="#cb19-67" aria-hidden="true" tabindex="-1"></a>db_ens_order <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">apply</span>(db_ens, <span class="dv">1</span>, <span class="at">FUN =</span> <span class="cf">function</span>(x) x[<span class="fu">order</span>(x)]))</span>
<span id="cb19-68"><a href="#cb19-68" aria-hidden="true" tabindex="-1"></a>db_ens <span class="ot">&lt;-</span> <span class="fu">unique</span>(db_ens_order) </span>
<span id="cb19-69"><a href="#cb19-69" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(db_ens_order)</span>
<span id="cb19-70"><a href="#cb19-70" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-71"><a href="#cb19-71" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert to data frame</span></span>
<span id="cb19-72"><a href="#cb19-72" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(db_ens) <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(db_ens)</span>
<span id="cb19-73"><a href="#cb19-73" aria-hidden="true" tabindex="-1"></a>db_ens <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(db_ens)</span>
<span id="cb19-74"><a href="#cb19-74" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-75"><a href="#cb19-75" aria-hidden="true" tabindex="-1"></a><span class="co"># Save as tab delimited file </span></span>
<span id="cb19-76"><a href="#cb19-76" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(<span class="at">x =</span> db_ens, </span>
<span id="cb19-77"><a href="#cb19-77" aria-hidden="true" tabindex="-1"></a>            <span class="at">file =</span> <span class="st">"./02_Datasets/CORUM/humanComplexes_EnsemblGeneID.txt"</span>, </span>
<span id="cb19-78"><a href="#cb19-78" aria-hidden="true" tabindex="-1"></a>            <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>, <span class="at">col.names =</span> <span class="cn">FALSE</span>, <span class="at">quote =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="partial-complexes" class="level3">
<h3 class="anchored" data-anchor-id="partial-complexes">Partial complexes</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># CORUM</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>db <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="at">file =</span> <span class="st">"./02_Datasets/CORUM/partialComplexes.txt"</span>, <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">header =</span> <span class="cn">TRUE</span>)</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Rename column names </span></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(db) <span class="ot">&lt;-</span> <span class="fu">str_replace_all</span>(<span class="at">string =</span> <span class="fu">colnames</span>(db), <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">."</span>, <span class="at">replacement =</span> <span class="st">"_"</span>)</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Reduce to human interactions</span></span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>db <span class="ot">&lt;-</span> db[<span class="fu">which</span>(db<span class="sc">$</span>Organism <span class="sc">==</span> <span class="st">"Human"</span>), ]</span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a><span class="co"># From complexes to pairs of interactions</span></span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a>output <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="cn">NA</span>, <span class="at">nrow =</span> <span class="dv">1</span>, <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">dimnames =</span> <span class="fu">list</span>(<span class="st">"RemoveMe"</span>, <span class="fu">c</span>(<span class="st">"Interactor_A"</span>, <span class="st">"Interactor_B"</span>)))</span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(db)){</span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a>  ci <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">str_split</span>(<span class="at">string =</span> db<span class="sc">$</span>subunits_UniProt_IDs_[i], <span class="at">pattern =</span> <span class="st">";"</span>))</span>
<span id="cb20-14"><a href="#cb20-14" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">length</span>(ci) <span class="sc">&gt;</span> <span class="dv">0</span>){</span>
<span id="cb20-15"><a href="#cb20-15" aria-hidden="true" tabindex="-1"></a>  output <span class="ot">&lt;-</span> <span class="fu">rbind</span>(output, <span class="fu">expand.grid</span>(<span class="st">"Interactor_A"</span> <span class="ot">=</span> ci, <span class="st">"Interactor_B"</span> <span class="ot">=</span> ci))  </span>
<span id="cb20-16"><a href="#cb20-16" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb20-17"><a href="#cb20-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rm</span>(ci)</span>
<span id="cb20-18"><a href="#cb20-18" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb20-19"><a href="#cb20-19" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(i)</span>
<span id="cb20-20"><a href="#cb20-20" aria-hidden="true" tabindex="-1"></a>output <span class="ot">&lt;-</span> output[<span class="sc">-</span><span class="dv">1</span>, ]</span>
<span id="cb20-21"><a href="#cb20-21" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(db)</span>
<span id="cb20-22"><a href="#cb20-22" aria-hidden="true" tabindex="-1"></a>db <span class="ot">&lt;-</span> output</span>
<span id="cb20-23"><a href="#cb20-23" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(output)</span>
<span id="cb20-24"><a href="#cb20-24" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(db) <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(db)</span>
<span id="cb20-25"><a href="#cb20-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-26"><a href="#cb20-26" aria-hidden="true" tabindex="-1"></a><span class="co"># Add ensembl annotations</span></span>
<span id="cb20-27"><a href="#cb20-27" aria-hidden="true" tabindex="-1"></a>db <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(db, </span>
<span id="cb20-28"><a href="#cb20-28" aria-hidden="true" tabindex="-1"></a>                 <span class="at">Ensembl_Gene_Interactor_A =</span> uniprot_to_ensembl<span class="sc">$</span>Ensembl_GeneID[<span class="fu">match</span>(<span class="at">x =</span> db<span class="sc">$</span>Interactor_A, <span class="at">table =</span> uniprot_to_ensembl<span class="sc">$</span>Uniprot_ID)],</span>
<span id="cb20-29"><a href="#cb20-29" aria-hidden="true" tabindex="-1"></a>                 <span class="at">Ensembl_Gene_Interactor_B =</span> uniprot_to_ensembl<span class="sc">$</span>Ensembl_GeneID[<span class="fu">match</span>(<span class="at">x =</span> db<span class="sc">$</span>Interactor_B, <span class="at">table =</span> uniprot_to_ensembl<span class="sc">$</span>Uniprot_ID)])</span>
<span id="cb20-30"><a href="#cb20-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-31"><a href="#cb20-31" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove interactions where either interactor cannot be mapped to Ensembl Gene IDs</span></span>
<span id="cb20-32"><a href="#cb20-32" aria-hidden="true" tabindex="-1"></a>nm <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">c</span>(<span class="fu">which</span>(<span class="fu">is.na</span>(db<span class="sc">$</span>Ensembl_Gene_Interactor_A) <span class="sc">==</span> <span class="cn">TRUE</span>),</span>
<span id="cb20-33"><a href="#cb20-33" aria-hidden="true" tabindex="-1"></a>               <span class="fu">which</span>(<span class="fu">is.na</span>(db<span class="sc">$</span>Ensembl_Gene_Interactor_B) <span class="sc">==</span> <span class="cn">TRUE</span>)))</span>
<span id="cb20-34"><a href="#cb20-34" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">length</span>(nm)<span class="sc">&gt;</span> <span class="dv">0</span>){ db <span class="ot">&lt;-</span> db[<span class="sc">-</span>nm, ]}</span>
<span id="cb20-35"><a href="#cb20-35" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(nm)</span>
<span id="cb20-36"><a href="#cb20-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-37"><a href="#cb20-37" aria-hidden="true" tabindex="-1"></a><span class="co"># Separate into interactions with a single and multiple Ensembl Gene IDs</span></span>
<span id="cb20-38"><a href="#cb20-38" aria-hidden="true" tabindex="-1"></a>db_m2m <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">c</span>(<span class="fu">which</span>(<span class="fu">str_detect</span>(<span class="at">string =</span> db<span class="sc">$</span>Ensembl_Gene_Interactor_A, <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">|"</span>) <span class="sc">==</span> <span class="cn">TRUE</span>),</span>
<span id="cb20-39"><a href="#cb20-39" aria-hidden="true" tabindex="-1"></a>                   <span class="fu">which</span>(<span class="fu">str_detect</span>(<span class="at">string =</span> db<span class="sc">$</span>Ensembl_Gene_Interactor_B, <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">|"</span>) <span class="sc">==</span> <span class="cn">TRUE</span>)))</span>
<span id="cb20-40"><a href="#cb20-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-41"><a href="#cb20-41" aria-hidden="true" tabindex="-1"></a>db_o2o <span class="ot">&lt;-</span> <span class="fu">setdiff</span>(<span class="at">x =</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(db), <span class="at">y =</span> db_m2m)</span>
<span id="cb20-42"><a href="#cb20-42" aria-hidden="true" tabindex="-1"></a>db_m2m <span class="ot">&lt;-</span> db[db_m2m, ]</span>
<span id="cb20-43"><a href="#cb20-43" aria-hidden="true" tabindex="-1"></a>db_o2o <span class="ot">&lt;-</span> db[db_o2o, ]</span>
<span id="cb20-44"><a href="#cb20-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-45"><a href="#cb20-45" aria-hidden="true" tabindex="-1"></a><span class="co"># Expand interactions with multiple ensembl gene IDs</span></span>
<span id="cb20-46"><a href="#cb20-46" aria-hidden="true" tabindex="-1"></a>output <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="cn">NA</span>, <span class="at">nrow =</span> <span class="dv">1</span>, <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">dimnames =</span> <span class="fu">list</span>(<span class="st">"ToRemove"</span>, <span class="fu">c</span>(<span class="st">"Ensembl_Gene_Interactor_A"</span>, <span class="st">"Ensembl_Gene_Interactor_B"</span>)))</span>
<span id="cb20-47"><a href="#cb20-47" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(db_m2m))){</span>
<span id="cb20-48"><a href="#cb20-48" aria-hidden="true" tabindex="-1"></a>  in_a <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">str_split</span>(<span class="at">string =</span> db_m2m<span class="sc">$</span>Ensembl_Gene_Interactor_A[i], <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">|</span><span class="sc">\\</span><span class="st">|"</span>))</span>
<span id="cb20-49"><a href="#cb20-49" aria-hidden="true" tabindex="-1"></a>  in_b <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">str_split</span>(<span class="at">string =</span> db_m2m<span class="sc">$</span>Ensembl_Gene_Interactor_B[i], <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">|</span><span class="sc">\\</span><span class="st">|"</span>))</span>
<span id="cb20-50"><a href="#cb20-50" aria-hidden="true" tabindex="-1"></a>  ab <span class="ot">&lt;-</span> <span class="fu">expand.grid</span>(<span class="st">"Ensembl_Gene_Interactor_A"</span> <span class="ot">=</span> in_a, <span class="st">"Ensembl_Gene_Interactor_B"</span> <span class="ot">=</span> in_b)</span>
<span id="cb20-51"><a href="#cb20-51" aria-hidden="true" tabindex="-1"></a>  output <span class="ot">&lt;-</span> <span class="fu">rbind</span>(output, ab)</span>
<span id="cb20-52"><a href="#cb20-52" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb20-53"><a href="#cb20-53" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(in_a, in_b, ab, i)</span>
<span id="cb20-54"><a href="#cb20-54" aria-hidden="true" tabindex="-1"></a>output <span class="ot">&lt;-</span> output[<span class="sc">-</span><span class="dv">1</span>, ]</span>
<span id="cb20-55"><a href="#cb20-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-56"><a href="#cb20-56" aria-hidden="true" tabindex="-1"></a><span class="co"># Combine interactions that map a single and multiple Ensembl Gene IDs </span></span>
<span id="cb20-57"><a href="#cb20-57" aria-hidden="true" tabindex="-1"></a>db_ens <span class="ot">&lt;-</span> <span class="fu">rbind</span>(db_o2o[, <span class="fu">c</span>(<span class="st">"Ensembl_Gene_Interactor_A"</span>, <span class="st">"Ensembl_Gene_Interactor_B"</span>)], </span>
<span id="cb20-58"><a href="#cb20-58" aria-hidden="true" tabindex="-1"></a>                output)</span>
<span id="cb20-59"><a href="#cb20-59" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(db, db_m2m, db_o2o, output)</span>
<span id="cb20-60"><a href="#cb20-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-61"><a href="#cb20-61" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove self interactions</span></span>
<span id="cb20-62"><a href="#cb20-62" aria-hidden="true" tabindex="-1"></a>mark_self <span class="ot">&lt;-</span> <span class="fu">apply</span>(db_ens, <span class="dv">1</span>, <span class="at">FUN =</span> <span class="cf">function</span>(x) <span class="fu">length</span>(<span class="fu">unique</span>(x)))</span>
<span id="cb20-63"><a href="#cb20-63" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">length</span>(<span class="fu">which</span>(mark_self <span class="sc">==</span> <span class="dv">1</span>))<span class="sc">&gt;</span> <span class="dv">0</span>){db_ens <span class="ot">&lt;-</span> db_ens[<span class="sc">-</span><span class="fu">which</span>(mark_self <span class="sc">==</span> <span class="dv">1</span>), ]}</span>
<span id="cb20-64"><a href="#cb20-64" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(mark_self)</span>
<span id="cb20-65"><a href="#cb20-65" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-66"><a href="#cb20-66" aria-hidden="true" tabindex="-1"></a><span class="co"># Re-order interactions to remove duplicates</span></span>
<span id="cb20-67"><a href="#cb20-67" aria-hidden="true" tabindex="-1"></a>db_ens_order <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">apply</span>(db_ens, <span class="dv">1</span>, <span class="at">FUN =</span> <span class="cf">function</span>(x) x[<span class="fu">order</span>(x)]))</span>
<span id="cb20-68"><a href="#cb20-68" aria-hidden="true" tabindex="-1"></a>db_ens <span class="ot">&lt;-</span> <span class="fu">unique</span>(db_ens_order) </span>
<span id="cb20-69"><a href="#cb20-69" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(db_ens_order)</span>
<span id="cb20-70"><a href="#cb20-70" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-71"><a href="#cb20-71" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert to data frame</span></span>
<span id="cb20-72"><a href="#cb20-72" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(db_ens) <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(db_ens)</span>
<span id="cb20-73"><a href="#cb20-73" aria-hidden="true" tabindex="-1"></a>db_ens <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(db_ens)</span>
<span id="cb20-74"><a href="#cb20-74" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-75"><a href="#cb20-75" aria-hidden="true" tabindex="-1"></a><span class="co"># Save as tab delimited file </span></span>
<span id="cb20-76"><a href="#cb20-76" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(<span class="at">x =</span> db_ens, </span>
<span id="cb20-77"><a href="#cb20-77" aria-hidden="true" tabindex="-1"></a>            <span class="at">file =</span> <span class="st">"./02_Datasets/CORUM/partialComplexes_EnsemblGeneID.txt"</span>, </span>
<span id="cb20-78"><a href="#cb20-78" aria-hidden="true" tabindex="-1"></a>            <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>, <span class="at">col.names =</span> <span class="cn">FALSE</span>, <span class="at">quote =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="hitpredict" class="level2">
<h2 class="anchored" data-anchor-id="hitpredict">HITPREDICT</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Download interactions - # August 2023 last update.</span></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a><span class="co"># You require to write an email to ashwini@combinatics.com to request access </span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Just send name, affiliation and brief description of use to get access link.</span></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a><span class="co"># tar zxvf ./02_Datasets/HITPREDICT/H_sapiens_interactions.txt.tgz</span></span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a><span class="co"># tar zxvf ./02_Datasets/HITPREDICT/HitPredict_interactions.txt.tgz</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co"># HitPredict latest</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>db <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="at">file =</span> <span class="st">"./02_Datasets/HITPREDICT/H_sapiens_interactions.txt"</span>, <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">header =</span> <span class="cn">TRUE</span>, <span class="at">skip =</span> <span class="dv">4</span>)</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Mapping through uniprot</span></span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>db <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(db, </span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>                 <span class="at">Ensembl_Gene_Interactor_A =</span> uniprot_to_ensembl<span class="sc">$</span>Ensembl_GeneID[<span class="fu">match</span>(<span class="at">x =</span> db<span class="sc">$</span>Uniprot1, <span class="at">table =</span> uniprot_to_ensembl<span class="sc">$</span>Uniprot_ID)],</span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a>                 <span class="at">Ensembl_Gene_Interactor_B =</span> uniprot_to_ensembl<span class="sc">$</span>Ensembl_GeneID[<span class="fu">match</span>(<span class="at">x =</span> db<span class="sc">$</span>Uniprot2, <span class="at">table =</span> uniprot_to_ensembl<span class="sc">$</span>Uniprot_ID)])</span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove interactions where either interactor cannot be mapped to Ensembl Gene IDs</span></span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a>nm <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">c</span>(<span class="fu">which</span>(<span class="fu">is.na</span>(db<span class="sc">$</span>Ensembl_Gene_Interactor_A) <span class="sc">==</span> <span class="cn">TRUE</span>),</span>
<span id="cb22-11"><a href="#cb22-11" aria-hidden="true" tabindex="-1"></a>               <span class="fu">which</span>(<span class="fu">is.na</span>(db<span class="sc">$</span>Ensembl_Gene_Interactor_B) <span class="sc">==</span> <span class="cn">TRUE</span>)))</span>
<span id="cb22-12"><a href="#cb22-12" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">length</span>(nm)<span class="sc">&gt;</span> <span class="dv">0</span>){ db <span class="ot">&lt;-</span> db[<span class="sc">-</span>nm, ]}</span>
<span id="cb22-13"><a href="#cb22-13" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(nm)</span>
<span id="cb22-14"><a href="#cb22-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-15"><a href="#cb22-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Separate into interactions with a single and multiple Ensembl Gene IDs</span></span>
<span id="cb22-16"><a href="#cb22-16" aria-hidden="true" tabindex="-1"></a>db_m2m <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">c</span>(<span class="fu">which</span>(<span class="fu">str_detect</span>(<span class="at">string =</span> db<span class="sc">$</span>Ensembl_Gene_Interactor_A, <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">|"</span>) <span class="sc">==</span> <span class="cn">TRUE</span>),</span>
<span id="cb22-17"><a href="#cb22-17" aria-hidden="true" tabindex="-1"></a>                   <span class="fu">which</span>(<span class="fu">str_detect</span>(<span class="at">string =</span> db<span class="sc">$</span>Ensembl_Gene_Interactor_B, <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">|"</span>) <span class="sc">==</span> <span class="cn">TRUE</span>)))</span>
<span id="cb22-18"><a href="#cb22-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-19"><a href="#cb22-19" aria-hidden="true" tabindex="-1"></a>db_o2o <span class="ot">&lt;-</span> <span class="fu">setdiff</span>(<span class="at">x =</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(db), <span class="at">y =</span> db_m2m)</span>
<span id="cb22-20"><a href="#cb22-20" aria-hidden="true" tabindex="-1"></a>db_m2m <span class="ot">&lt;-</span> db[db_m2m, ]</span>
<span id="cb22-21"><a href="#cb22-21" aria-hidden="true" tabindex="-1"></a>db_o2o <span class="ot">&lt;-</span> db[db_o2o, ]</span>
<span id="cb22-22"><a href="#cb22-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-23"><a href="#cb22-23" aria-hidden="true" tabindex="-1"></a><span class="co"># Expand interactions with multiple ensembl gene IDs</span></span>
<span id="cb22-24"><a href="#cb22-24" aria-hidden="true" tabindex="-1"></a>output <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="cn">NA</span>, <span class="at">nrow =</span> <span class="dv">1</span>, <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">dimnames =</span> <span class="fu">list</span>(<span class="st">"ToRemove"</span>, <span class="fu">c</span>(<span class="st">"Ensembl_Gene_Interactor_A"</span>, <span class="st">"Ensembl_Gene_Interactor_B"</span>)))</span>
<span id="cb22-25"><a href="#cb22-25" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(db_m2m))){</span>
<span id="cb22-26"><a href="#cb22-26" aria-hidden="true" tabindex="-1"></a>  in_a <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">str_split</span>(<span class="at">string =</span> db_m2m<span class="sc">$</span>Ensembl_Gene_Interactor_A[i], <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">|</span><span class="sc">\\</span><span class="st">|"</span>))</span>
<span id="cb22-27"><a href="#cb22-27" aria-hidden="true" tabindex="-1"></a>  in_b <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">str_split</span>(<span class="at">string =</span> db_m2m<span class="sc">$</span>Ensembl_Gene_Interactor_B[i], <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">|</span><span class="sc">\\</span><span class="st">|"</span>))</span>
<span id="cb22-28"><a href="#cb22-28" aria-hidden="true" tabindex="-1"></a>  ab <span class="ot">&lt;-</span> <span class="fu">expand.grid</span>(<span class="st">"Ensembl_Gene_Interactor_A"</span> <span class="ot">=</span> in_a, <span class="st">"Ensembl_Gene_Interactor_B"</span> <span class="ot">=</span> in_b)</span>
<span id="cb22-29"><a href="#cb22-29" aria-hidden="true" tabindex="-1"></a>  output <span class="ot">&lt;-</span> <span class="fu">rbind</span>(output, ab)</span>
<span id="cb22-30"><a href="#cb22-30" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb22-31"><a href="#cb22-31" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(in_a, in_b, ab, i)</span>
<span id="cb22-32"><a href="#cb22-32" aria-hidden="true" tabindex="-1"></a>output <span class="ot">&lt;-</span> output[<span class="sc">-</span><span class="dv">1</span>, ]</span>
<span id="cb22-33"><a href="#cb22-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-34"><a href="#cb22-34" aria-hidden="true" tabindex="-1"></a><span class="co"># Combine interactions that map a single and multiple Ensembl Gene IDs </span></span>
<span id="cb22-35"><a href="#cb22-35" aria-hidden="true" tabindex="-1"></a>db_ens <span class="ot">&lt;-</span> <span class="fu">rbind</span>(db_o2o[, <span class="fu">c</span>(<span class="st">"Ensembl_Gene_Interactor_A"</span>, <span class="st">"Ensembl_Gene_Interactor_B"</span>)], </span>
<span id="cb22-36"><a href="#cb22-36" aria-hidden="true" tabindex="-1"></a>                output)</span>
<span id="cb22-37"><a href="#cb22-37" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(db, db_m2m, db_o2o, output)</span>
<span id="cb22-38"><a href="#cb22-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-39"><a href="#cb22-39" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove self interactions</span></span>
<span id="cb22-40"><a href="#cb22-40" aria-hidden="true" tabindex="-1"></a>mark_self <span class="ot">&lt;-</span> <span class="fu">apply</span>(db_ens, <span class="dv">1</span>, <span class="at">FUN =</span> <span class="cf">function</span>(x) <span class="fu">length</span>(<span class="fu">unique</span>(x)))</span>
<span id="cb22-41"><a href="#cb22-41" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">length</span>(<span class="fu">which</span>(mark_self <span class="sc">==</span> <span class="dv">1</span>))<span class="sc">&gt;</span> <span class="dv">0</span>){db_ens <span class="ot">&lt;-</span> db_ens[<span class="sc">-</span><span class="fu">which</span>(mark_self <span class="sc">==</span> <span class="dv">1</span>), ]}</span>
<span id="cb22-42"><a href="#cb22-42" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(mark_self)</span>
<span id="cb22-43"><a href="#cb22-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-44"><a href="#cb22-44" aria-hidden="true" tabindex="-1"></a><span class="co"># Re-order interactions to remove duplicates</span></span>
<span id="cb22-45"><a href="#cb22-45" aria-hidden="true" tabindex="-1"></a>db_ens_order <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">apply</span>(db_ens, <span class="dv">1</span>, <span class="at">FUN =</span> <span class="cf">function</span>(x) x[<span class="fu">order</span>(x)]))</span>
<span id="cb22-46"><a href="#cb22-46" aria-hidden="true" tabindex="-1"></a>db_ens <span class="ot">&lt;-</span> <span class="fu">unique</span>(db_ens_order) </span>
<span id="cb22-47"><a href="#cb22-47" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(db_ens_order)</span>
<span id="cb22-48"><a href="#cb22-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-49"><a href="#cb22-49" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert to data frame</span></span>
<span id="cb22-50"><a href="#cb22-50" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(db_ens) <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(db_ens)</span>
<span id="cb22-51"><a href="#cb22-51" aria-hidden="true" tabindex="-1"></a>db_ens <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(db_ens)</span>
<span id="cb22-52"><a href="#cb22-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-53"><a href="#cb22-53" aria-hidden="true" tabindex="-1"></a><span class="co"># Save as tab delimited file </span></span>
<span id="cb22-54"><a href="#cb22-54" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(<span class="at">x =</span> db_ens, </span>
<span id="cb22-55"><a href="#cb22-55" aria-hidden="true" tabindex="-1"></a>            <span class="at">file =</span> <span class="st">"./02_Datasets/HITPREDICT/H_sapiens_interactions_EnsemblGeneID.txt"</span>, </span>
<span id="cb22-56"><a href="#cb22-56" aria-hidden="true" tabindex="-1"></a>            <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>, <span class="at">col.names =</span> <span class="cn">FALSE</span>, <span class="at">quote =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="huri" class="level2">
<h2 class="anchored" data-anchor-id="huri">HURI</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="at">path =</span> <span class="st">"./02_Datasets/HURI/"</span>)</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(<span class="at">url =</span> <span class="st">"http://www.interactome-atlas.org/data/HuRI.tsv"</span>,</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>              <span class="at">destfile =</span> <span class="st">"./02_Datasets/HURI/HuRI.tsv"</span>)</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(<span class="at">url =</span> <span class="st">"http://www.interactome-atlas.org/data/HI-union.tsv"</span>, </span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>              <span class="at">destfile =</span> <span class="st">"./02_Datasets/HURI/HI-union.tsv"</span>)</span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(<span class="at">url =</span> <span class="st">"http://www.interactome-atlas.org/data/Test_space_screens-19.tsv"</span>, </span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a>              <span class="at">destfile =</span> <span class="st">"./02_Datasets/HURI/Test_space_screens-19.tsv"</span>)</span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(<span class="at">url =</span> <span class="st">"http://www.interactome-atlas.org/data/Lit-BM.tsv"</span>, </span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a>              <span class="at">destfile =</span> <span class="st">"./02_Datasets/HURI/Lit-BM.tsv"</span>)</span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-14"><a href="#cb23-14" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(<span class="at">url =</span> <span class="st">"http://www.interactome-atlas.org/data/Yang-16.tsv"</span>, </span>
<span id="cb23-15"><a href="#cb23-15" aria-hidden="true" tabindex="-1"></a>              <span class="at">destfile =</span> <span class="st">"./02_Datasets/HURI/Yang-16.tsv"</span>)</span>
<span id="cb23-16"><a href="#cb23-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-17"><a href="#cb23-17" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(<span class="at">url =</span> <span class="st">"http://www.interactome-atlas.org/data/HI-II-14.tsv"</span>, </span>
<span id="cb23-18"><a href="#cb23-18" aria-hidden="true" tabindex="-1"></a>              <span class="at">destfile =</span> <span class="st">"./02_Datasets/HURI/HI-II-14.tsv"</span>)</span>
<span id="cb23-19"><a href="#cb23-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-20"><a href="#cb23-20" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(<span class="at">url =</span> <span class="st">"http://www.interactome-atlas.org/data/Venkatesan-09.tsv"</span>, </span>
<span id="cb23-21"><a href="#cb23-21" aria-hidden="true" tabindex="-1"></a>              <span class="at">destfile =</span> <span class="st">"./02_Datasets/HURI/Venkatesan-09.tsv"</span>)</span>
<span id="cb23-22"><a href="#cb23-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-23"><a href="#cb23-23" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(<span class="at">url =</span> <span class="st">"http://www.interactome-atlas.org/data/H-I-05.tsv"</span>,</span>
<span id="cb23-24"><a href="#cb23-24" aria-hidden="true" tabindex="-1"></a>              <span class="at">destfile =</span> <span class="st">"./02_Datasets/HURI/H-I-05.tsv"</span>)</span>
<span id="cb23-25"><a href="#cb23-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-26"><a href="#cb23-26" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(<span class="at">url =</span> <span class="st">"http://www.interactome-atlas.org/data/Yu-11.tsv"</span>,</span>
<span id="cb23-27"><a href="#cb23-27" aria-hidden="true" tabindex="-1"></a>              <span class="at">destfile =</span> <span class="st">"./02_Datasets/HURI/Yu-11.tsv"</span>)</span>
<span id="cb23-28"><a href="#cb23-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-29"><a href="#cb23-29" aria-hidden="true" tabindex="-1"></a>fn <span class="ot">&lt;-</span> <span class="fu">list.files</span>(<span class="at">path =</span> <span class="st">"./02_Datasets/HURI/"</span>, <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(fn))){</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># HuRI</span></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>  db_ens <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="at">file =</span> fn[i], <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">header =</span> <span class="cn">FALSE</span>)</span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Remove self interactions</span></span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>  mark_self <span class="ot">&lt;-</span> <span class="fu">apply</span>(db_ens, <span class="dv">1</span>, <span class="at">FUN =</span> <span class="cf">function</span>(x) <span class="fu">length</span>(<span class="fu">unique</span>(x)))</span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">length</span>(<span class="fu">which</span>(mark_self <span class="sc">==</span> <span class="dv">1</span>))<span class="sc">&gt;</span> <span class="dv">0</span>){db_ens <span class="ot">&lt;-</span> db_ens[<span class="sc">-</span><span class="fu">which</span>(mark_self <span class="sc">==</span> <span class="dv">1</span>), ]}</span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rm</span>(mark_self)</span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Re-order interactions to remove duplicates</span></span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a>  db_ens_order <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">apply</span>(db_ens, <span class="dv">1</span>, <span class="at">FUN =</span> <span class="cf">function</span>(x) x[<span class="fu">order</span>(x)]))</span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true" tabindex="-1"></a>  db_ens <span class="ot">&lt;-</span> <span class="fu">unique</span>(db_ens_order) </span>
<span id="cb24-14"><a href="#cb24-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rm</span>(db_ens_order)</span>
<span id="cb24-15"><a href="#cb24-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-16"><a href="#cb24-16" aria-hidden="true" tabindex="-1"></a>  out_name <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="fu">str_replace</span>(<span class="at">string =</span> fn[i], </span>
<span id="cb24-17"><a href="#cb24-17" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">pattern =</span> <span class="st">".tsv"</span>, </span>
<span id="cb24-18"><a href="#cb24-18" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">replacement =</span> <span class="st">""</span>), </span>
<span id="cb24-19"><a href="#cb24-19" aria-hidden="true" tabindex="-1"></a>                     <span class="st">"_EnsemblGeneID.txt"</span>, <span class="at">collapse =</span> <span class="st">""</span>)</span>
<span id="cb24-20"><a href="#cb24-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-21"><a href="#cb24-21" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Save as tab delimited file </span></span>
<span id="cb24-22"><a href="#cb24-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">write.table</span>(<span class="at">x =</span> db_ens, </span>
<span id="cb24-23"><a href="#cb24-23" aria-hidden="true" tabindex="-1"></a>              <span class="at">file =</span> out_name, </span>
<span id="cb24-24"><a href="#cb24-24" aria-hidden="true" tabindex="-1"></a>              <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>, <span class="at">col.names =</span> <span class="cn">FALSE</span>, <span class="at">quote =</span> <span class="cn">FALSE</span>)</span>
<span id="cb24-25"><a href="#cb24-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rm</span>(db_ens, out_name)</span>
<span id="cb24-26"><a href="#cb24-26" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb24-27"><a href="#cb24-27" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(fn, i)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="intact" class="level2">
<h2 class="anchored" data-anchor-id="intact">INTACT</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Data download</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="at">path =</span> <span class="st">"./02_Datasets/INTACT/"</span>)</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(<span class="at">url =</span> <span class="st">"https://ftp.ebi.ac.uk/pub/databases/intact/current/psimitab/intact.zip"</span>,</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>              <span class="at">destfile =</span> <span class="st">"./02_Datasets/INTACT/intact.zip"</span>) </span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a><span class="fu">unzip</span>(<span class="at">zipfile =</span> <span class="st">"./02_Datasets/INTACT/intact.zip"</span>, <span class="at">exdir =</span> <span class="st">"./02_Datasets/INTACT/"</span>)</span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a><span class="fu">unlink</span>(<span class="at">x =</span> <span class="st">"./02_Datasets/INTACT/intact.zip"</span>)</span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a><span class="co"># INTACT files to system variables </span></span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a><span class="fu">Sys.setenv</span>(<span class="at">INTACT =</span> <span class="st">"./02_Datasets/INTACT/intact.txt"</span>)</span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a><span class="fu">Sys.setenv</span>(<span class="at">INTACT_TMP =</span> <span class="st">"./02_Datasets/INTACT/intact_tmp.txt"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Parse file to remove ' and # from file</span></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a><span class="fu">sed</span> <span class="st">"s/'//g"</span> <span class="va">$INTACT</span> <span class="op">&gt;</span> <span class="va">$INTACT_TMP</span></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a><span class="fu">sed</span> <span class="st">"s/#//g"</span> <span class="va">$INTACT_TMP</span> <span class="op">&gt;</span> <span class="va">$INTACT</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove intermediate files</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="fu">unlink</span>(<span class="at">x =</span> <span class="st">"./02_Datasets/INTACT/intact_tmp.txt"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Intact db</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>db <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="at">file =</span> <span class="st">"./02_Datasets/INTACT/intact.txt"</span>, <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">quote =</span> <span class="st">""</span>, <span class="at">header =</span> <span class="cn">TRUE</span>)</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Replace "." from colnames </span></span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(db) <span class="ot">&lt;-</span> <span class="fu">str_replace_all</span>(<span class="at">string =</span> <span class="fu">colnames</span>(db), <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">.</span><span class="sc">\\</span><span class="st">."</span>, <span class="at">replacement =</span> <span class="st">"."</span>)</span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(db) <span class="ot">&lt;-</span> <span class="fu">str_replace_all</span>(<span class="at">string =</span> <span class="fu">colnames</span>(db), <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">."</span>, <span class="at">replacement =</span> <span class="st">"_"</span>)</span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Only within human interactions</span></span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a>hsap <span class="ot">&lt;-</span> <span class="fu">intersect</span>(<span class="fu">which</span>(<span class="fu">str_detect</span>(<span class="at">string =</span> db<span class="sc">$</span>Taxid_interactor_A, <span class="at">pattern =</span> <span class="st">"taxid:9606"</span>) <span class="sc">==</span> <span class="cn">TRUE</span>), </span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a>                  <span class="fu">which</span>(<span class="fu">str_detect</span>(<span class="at">string =</span> db<span class="sc">$</span>Taxid_interactor_B, <span class="at">pattern =</span> <span class="st">"taxid:9606"</span>) <span class="sc">==</span> <span class="cn">TRUE</span>))</span>
<span id="cb28-11"><a href="#cb28-11" aria-hidden="true" tabindex="-1"></a>db <span class="ot">&lt;-</span> db[hsap, ]</span>
<span id="cb28-12"><a href="#cb28-12" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(hsap)</span>
<span id="cb28-13"><a href="#cb28-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-14"><a href="#cb28-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Mapping through uniprot</span></span>
<span id="cb28-15"><a href="#cb28-15" aria-hidden="true" tabindex="-1"></a>uni <span class="ot">&lt;-</span> <span class="fu">intersect</span>(<span class="fu">which</span>(<span class="fu">str_detect</span>(<span class="at">string =</span> db<span class="sc">$</span>ID_s_interactor_A, <span class="at">pattern =</span> <span class="st">"uniprotkb"</span>) <span class="sc">==</span> <span class="cn">TRUE</span>),</span>
<span id="cb28-16"><a href="#cb28-16" aria-hidden="true" tabindex="-1"></a>                 <span class="fu">which</span>(<span class="fu">str_detect</span>(<span class="at">string =</span> db<span class="sc">$</span>ID_s_interactor_B, <span class="at">pattern =</span> <span class="st">"uniprotkb"</span>) <span class="sc">==</span> <span class="cn">TRUE</span>))</span>
<span id="cb28-17"><a href="#cb28-17" aria-hidden="true" tabindex="-1"></a>db <span class="ot">&lt;-</span> db[uni, ]</span>
<span id="cb28-18"><a href="#cb28-18" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(uni)</span>
<span id="cb28-19"><a href="#cb28-19" aria-hidden="true" tabindex="-1"></a>db<span class="sc">$</span>ID_s_interactor_A <span class="ot">&lt;-</span> <span class="fu">str_replace</span>(<span class="at">string =</span> db<span class="sc">$</span>ID_s_interactor_A, <span class="at">pattern =</span> <span class="st">"uniprotkb:"</span>, <span class="at">replacement =</span> <span class="st">""</span>)</span>
<span id="cb28-20"><a href="#cb28-20" aria-hidden="true" tabindex="-1"></a>db<span class="sc">$</span>ID_s_interactor_B <span class="ot">&lt;-</span> <span class="fu">str_replace</span>(<span class="at">string =</span> db<span class="sc">$</span>ID_s_interactor_B, <span class="at">pattern =</span> <span class="st">"uniprotkb:"</span>, <span class="at">replacement =</span> <span class="st">""</span>)</span>
<span id="cb28-21"><a href="#cb28-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-22"><a href="#cb28-22" aria-hidden="true" tabindex="-1"></a>db <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(db, </span>
<span id="cb28-23"><a href="#cb28-23" aria-hidden="true" tabindex="-1"></a>                 <span class="at">Ensembl_Gene_Interactor_A =</span> uniprot_to_ensembl<span class="sc">$</span>Ensembl_GeneID[<span class="fu">match</span>(<span class="at">x =</span> db<span class="sc">$</span>ID_s_interactor_A, <span class="at">table =</span> uniprot_to_ensembl<span class="sc">$</span>Uniprot_ID)],</span>
<span id="cb28-24"><a href="#cb28-24" aria-hidden="true" tabindex="-1"></a>                 <span class="at">Ensembl_Gene_Interactor_B =</span> uniprot_to_ensembl<span class="sc">$</span>Ensembl_GeneID[<span class="fu">match</span>(<span class="at">x =</span> db<span class="sc">$</span>ID_s_interactor_B, <span class="at">table =</span> uniprot_to_ensembl<span class="sc">$</span>Uniprot_ID)])</span>
<span id="cb28-25"><a href="#cb28-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-26"><a href="#cb28-26" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove interactions where either interactor cannot be mapped to Ensembl Gene IDs</span></span>
<span id="cb28-27"><a href="#cb28-27" aria-hidden="true" tabindex="-1"></a>nm <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">c</span>(<span class="fu">which</span>(<span class="fu">is.na</span>(db<span class="sc">$</span>Ensembl_Gene_Interactor_A) <span class="sc">==</span> <span class="cn">TRUE</span>),</span>
<span id="cb28-28"><a href="#cb28-28" aria-hidden="true" tabindex="-1"></a>               <span class="fu">which</span>(<span class="fu">is.na</span>(db<span class="sc">$</span>Ensembl_Gene_Interactor_B) <span class="sc">==</span> <span class="cn">TRUE</span>)))</span>
<span id="cb28-29"><a href="#cb28-29" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">length</span>(nm)<span class="sc">&gt;</span> <span class="dv">0</span>){ db <span class="ot">&lt;-</span> db[<span class="sc">-</span>nm, ]}</span>
<span id="cb28-30"><a href="#cb28-30" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(nm)</span>
<span id="cb28-31"><a href="#cb28-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-32"><a href="#cb28-32" aria-hidden="true" tabindex="-1"></a><span class="co"># Separate into interactions with a single and multiple Ensembl Gene IDs</span></span>
<span id="cb28-33"><a href="#cb28-33" aria-hidden="true" tabindex="-1"></a>db_m2m <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">c</span>(<span class="fu">which</span>(<span class="fu">str_detect</span>(<span class="at">string =</span> db<span class="sc">$</span>Ensembl_Gene_Interactor_A, <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">|"</span>) <span class="sc">==</span> <span class="cn">TRUE</span>),</span>
<span id="cb28-34"><a href="#cb28-34" aria-hidden="true" tabindex="-1"></a>                   <span class="fu">which</span>(<span class="fu">str_detect</span>(<span class="at">string =</span> db<span class="sc">$</span>Ensembl_Gene_Interactor_B, <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">|"</span>) <span class="sc">==</span> <span class="cn">TRUE</span>)))</span>
<span id="cb28-35"><a href="#cb28-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-36"><a href="#cb28-36" aria-hidden="true" tabindex="-1"></a>db_o2o <span class="ot">&lt;-</span> <span class="fu">setdiff</span>(<span class="at">x =</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(db), <span class="at">y =</span> db_m2m)</span>
<span id="cb28-37"><a href="#cb28-37" aria-hidden="true" tabindex="-1"></a>db_m2m <span class="ot">&lt;-</span> db[db_m2m, ]</span>
<span id="cb28-38"><a href="#cb28-38" aria-hidden="true" tabindex="-1"></a>db_o2o <span class="ot">&lt;-</span> db[db_o2o, ]</span>
<span id="cb28-39"><a href="#cb28-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-40"><a href="#cb28-40" aria-hidden="true" tabindex="-1"></a><span class="co"># Expand interactions with multiple ensembl gene IDs</span></span>
<span id="cb28-41"><a href="#cb28-41" aria-hidden="true" tabindex="-1"></a>output <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="cn">NA</span>, <span class="at">nrow =</span> <span class="dv">1</span>, <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">dimnames =</span> <span class="fu">list</span>(<span class="st">"ToRemove"</span>, <span class="fu">c</span>(<span class="st">"Ensembl_Gene_Interactor_A"</span>, <span class="st">"Ensembl_Gene_Interactor_B"</span>)))</span>
<span id="cb28-42"><a href="#cb28-42" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(db_m2m))){</span>
<span id="cb28-43"><a href="#cb28-43" aria-hidden="true" tabindex="-1"></a>  in_a <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">str_split</span>(<span class="at">string =</span> db_m2m<span class="sc">$</span>Ensembl_Gene_Interactor_A[i], <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">|</span><span class="sc">\\</span><span class="st">|"</span>))</span>
<span id="cb28-44"><a href="#cb28-44" aria-hidden="true" tabindex="-1"></a>  in_b <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">str_split</span>(<span class="at">string =</span> db_m2m<span class="sc">$</span>Ensembl_Gene_Interactor_B[i], <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">|</span><span class="sc">\\</span><span class="st">|"</span>))</span>
<span id="cb28-45"><a href="#cb28-45" aria-hidden="true" tabindex="-1"></a>  ab <span class="ot">&lt;-</span> <span class="fu">expand.grid</span>(<span class="st">"Ensembl_Gene_Interactor_A"</span> <span class="ot">=</span> in_a, <span class="st">"Ensembl_Gene_Interactor_B"</span> <span class="ot">=</span> in_b)</span>
<span id="cb28-46"><a href="#cb28-46" aria-hidden="true" tabindex="-1"></a>  output <span class="ot">&lt;-</span> <span class="fu">rbind</span>(output, ab)</span>
<span id="cb28-47"><a href="#cb28-47" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb28-48"><a href="#cb28-48" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(in_a, in_b, ab, i)</span>
<span id="cb28-49"><a href="#cb28-49" aria-hidden="true" tabindex="-1"></a>output <span class="ot">&lt;-</span> output[<span class="sc">-</span><span class="dv">1</span>, ]</span>
<span id="cb28-50"><a href="#cb28-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-51"><a href="#cb28-51" aria-hidden="true" tabindex="-1"></a><span class="co"># Combine interactions that map a single and multiple Ensembl Gene IDs </span></span>
<span id="cb28-52"><a href="#cb28-52" aria-hidden="true" tabindex="-1"></a>db_ens <span class="ot">&lt;-</span> <span class="fu">rbind</span>(db_o2o[, <span class="fu">c</span>(<span class="st">"Ensembl_Gene_Interactor_A"</span>, <span class="st">"Ensembl_Gene_Interactor_B"</span>)], </span>
<span id="cb28-53"><a href="#cb28-53" aria-hidden="true" tabindex="-1"></a>                output)</span>
<span id="cb28-54"><a href="#cb28-54" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(db, db_m2m, db_o2o, output)</span>
<span id="cb28-55"><a href="#cb28-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-56"><a href="#cb28-56" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove self interactions</span></span>
<span id="cb28-57"><a href="#cb28-57" aria-hidden="true" tabindex="-1"></a>mark_self <span class="ot">&lt;-</span> <span class="fu">apply</span>(db_ens, <span class="dv">1</span>, <span class="at">FUN =</span> <span class="cf">function</span>(x) <span class="fu">length</span>(<span class="fu">unique</span>(x)))</span>
<span id="cb28-58"><a href="#cb28-58" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">length</span>(<span class="fu">which</span>(mark_self <span class="sc">==</span> <span class="dv">1</span>))<span class="sc">&gt;</span> <span class="dv">0</span>){db_ens <span class="ot">&lt;-</span> db_ens[<span class="sc">-</span><span class="fu">which</span>(mark_self <span class="sc">==</span> <span class="dv">1</span>), ]}</span>
<span id="cb28-59"><a href="#cb28-59" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(mark_self)</span>
<span id="cb28-60"><a href="#cb28-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-61"><a href="#cb28-61" aria-hidden="true" tabindex="-1"></a><span class="co"># Re-order interactions to remove duplicates</span></span>
<span id="cb28-62"><a href="#cb28-62" aria-hidden="true" tabindex="-1"></a>db_ens_order <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">apply</span>(db_ens, <span class="dv">1</span>, <span class="at">FUN =</span> <span class="cf">function</span>(x) x[<span class="fu">order</span>(x)]))</span>
<span id="cb28-63"><a href="#cb28-63" aria-hidden="true" tabindex="-1"></a>db_ens <span class="ot">&lt;-</span> <span class="fu">unique</span>(db_ens_order) </span>
<span id="cb28-64"><a href="#cb28-64" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(db_ens_order)</span>
<span id="cb28-65"><a href="#cb28-65" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-66"><a href="#cb28-66" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert to data frame</span></span>
<span id="cb28-67"><a href="#cb28-67" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(db_ens) <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(db_ens)</span>
<span id="cb28-68"><a href="#cb28-68" aria-hidden="true" tabindex="-1"></a>db_ens <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(db_ens)</span>
<span id="cb28-69"><a href="#cb28-69" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-70"><a href="#cb28-70" aria-hidden="true" tabindex="-1"></a><span class="co"># Save as tab delimited file </span></span>
<span id="cb28-71"><a href="#cb28-71" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(<span class="at">x =</span> db_ens, </span>
<span id="cb28-72"><a href="#cb28-72" aria-hidden="true" tabindex="-1"></a>            <span class="at">file =</span> <span class="st">"./02_Datasets/INTACT/intact_EnsemblGeneID.txt"</span>, </span>
<span id="cb28-73"><a href="#cb28-73" aria-hidden="true" tabindex="-1"></a>            <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>, <span class="at">col.names =</span> <span class="cn">FALSE</span>, <span class="at">quote =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="mint" class="level2">
<h2 class="anchored" data-anchor-id="mint">MINT</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Data download</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="at">path =</span> <span class="st">"./02_Datasets/MINT/"</span>)</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(<span class="at">url =</span> <span class="st">"http://www.ebi.ac.uk/Tools/webservices/psicquic/mint/webservices/current/search/query/species:human"</span>,</span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>              <span class="at">destfile =</span> <span class="st">"./02_Datasets/MINT/species_human.txt"</span>)</span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a><span class="fu">Sys.setenv</span>(<span class="at">MINT =</span> <span class="st">"./02_Datasets/MINT/species_human.txt"</span>)</span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a><span class="fu">Sys.setenv</span>(<span class="at">MINT_4R =</span> <span class="st">"./02_Datasets/MINT/species_human_tmp.txt"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Parse file to remove ' and # from file</span></span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a><span class="fu">sed</span> <span class="st">"s/'//g"</span> <span class="va">$MINT</span> <span class="op">&gt;</span> <span class="va">$MINT_4R</span></span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a><span class="fu">sed</span> <span class="st">"s/#//g"</span> <span class="va">$MINT_4R</span> <span class="op">&gt;</span> <span class="va">$MINT</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove temporal file</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="fu">unlink</span>(<span class="at">x =</span> <span class="st">"./02_Datasets/MINT/species_human_tmp.txt"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Mint</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>db <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(<span class="st">"./02_Datasets/MINT/species_human.txt"</span>, <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">quote =</span> <span class="cn">NULL</span>, <span class="at">header =</span> <span class="cn">FALSE</span>, <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>)</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Rename a few columns</span></span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(db)[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>] <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Interactor_A"</span>, <span class="st">"Interactor_B"</span>)</span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(db)[<span class="dv">10</span><span class="sc">:</span><span class="dv">11</span>] <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Taxid_interactor_A"</span>, <span class="st">"Taxid_interactor_B"</span>)</span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Only within human interactions</span></span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a>hsap <span class="ot">&lt;-</span> <span class="fu">intersect</span>(<span class="fu">which</span>(<span class="fu">str_detect</span>(<span class="at">string =</span> db<span class="sc">$</span>Taxid_interactor_A, <span class="at">pattern =</span> <span class="st">"taxid:9606"</span>) <span class="sc">==</span> <span class="cn">TRUE</span>), </span>
<span id="cb32-10"><a href="#cb32-10" aria-hidden="true" tabindex="-1"></a>                  <span class="fu">which</span>(<span class="fu">str_detect</span>(<span class="at">string =</span> db<span class="sc">$</span>Taxid_interactor_B, <span class="at">pattern =</span> <span class="st">"taxid:9606"</span>) <span class="sc">==</span> <span class="cn">TRUE</span>))</span>
<span id="cb32-11"><a href="#cb32-11" aria-hidden="true" tabindex="-1"></a>db <span class="ot">&lt;-</span> db[hsap, ]</span>
<span id="cb32-12"><a href="#cb32-12" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(hsap)</span>
<span id="cb32-13"><a href="#cb32-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-14"><a href="#cb32-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Mapping through uniprot</span></span>
<span id="cb32-15"><a href="#cb32-15" aria-hidden="true" tabindex="-1"></a>uni <span class="ot">&lt;-</span> <span class="fu">intersect</span>(<span class="fu">which</span>(<span class="fu">str_detect</span>(<span class="at">string =</span> db<span class="sc">$</span>Interactor_A, <span class="at">pattern =</span> <span class="st">"uniprotkb"</span>) <span class="sc">==</span> <span class="cn">TRUE</span>),</span>
<span id="cb32-16"><a href="#cb32-16" aria-hidden="true" tabindex="-1"></a>                 <span class="fu">which</span>(<span class="fu">str_detect</span>(<span class="at">string =</span> db<span class="sc">$</span>Interactor_B, <span class="at">pattern =</span> <span class="st">"uniprotkb"</span>) <span class="sc">==</span> <span class="cn">TRUE</span>))</span>
<span id="cb32-17"><a href="#cb32-17" aria-hidden="true" tabindex="-1"></a>db <span class="ot">&lt;-</span> db[uni, ]</span>
<span id="cb32-18"><a href="#cb32-18" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(uni)</span>
<span id="cb32-19"><a href="#cb32-19" aria-hidden="true" tabindex="-1"></a>db<span class="sc">$</span>Interactor_A <span class="ot">&lt;-</span> <span class="fu">str_replace</span>(<span class="at">string =</span> db<span class="sc">$</span>Interactor_A, <span class="at">pattern =</span> <span class="st">"uniprotkb:"</span>, <span class="at">replacement =</span> <span class="st">""</span>)</span>
<span id="cb32-20"><a href="#cb32-20" aria-hidden="true" tabindex="-1"></a>db<span class="sc">$</span>Interactor_B <span class="ot">&lt;-</span> <span class="fu">str_replace</span>(<span class="at">string =</span> db<span class="sc">$</span>Interactor_B, <span class="at">pattern =</span> <span class="st">"uniprotkb:"</span>, <span class="at">replacement =</span> <span class="st">""</span>)</span>
<span id="cb32-21"><a href="#cb32-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-22"><a href="#cb32-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Mapping through uniprot</span></span>
<span id="cb32-23"><a href="#cb32-23" aria-hidden="true" tabindex="-1"></a>db <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(db, </span>
<span id="cb32-24"><a href="#cb32-24" aria-hidden="true" tabindex="-1"></a>                 <span class="at">Ensembl_Gene_Interactor_A =</span> uniprot_to_ensembl<span class="sc">$</span>Ensembl_GeneID[<span class="fu">match</span>(<span class="at">x =</span> db<span class="sc">$</span>Interactor_A, <span class="at">table =</span> uniprot_to_ensembl<span class="sc">$</span>Uniprot_ID)],</span>
<span id="cb32-25"><a href="#cb32-25" aria-hidden="true" tabindex="-1"></a>                 <span class="at">Ensembl_Gene_Interactor_B =</span> uniprot_to_ensembl<span class="sc">$</span>Ensembl_GeneID[<span class="fu">match</span>(<span class="at">x =</span> db<span class="sc">$</span>Interactor_B, <span class="at">table =</span> uniprot_to_ensembl<span class="sc">$</span>Uniprot_ID)])</span>
<span id="cb32-26"><a href="#cb32-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-27"><a href="#cb32-27" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove interactions where either interactor cannot be mapped to Ensembl Gene IDs</span></span>
<span id="cb32-28"><a href="#cb32-28" aria-hidden="true" tabindex="-1"></a>nm <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">c</span>(<span class="fu">which</span>(<span class="fu">is.na</span>(db<span class="sc">$</span>Ensembl_Gene_Interactor_A) <span class="sc">==</span> <span class="cn">TRUE</span>),</span>
<span id="cb32-29"><a href="#cb32-29" aria-hidden="true" tabindex="-1"></a>               <span class="fu">which</span>(<span class="fu">is.na</span>(db<span class="sc">$</span>Ensembl_Gene_Interactor_B) <span class="sc">==</span> <span class="cn">TRUE</span>)))</span>
<span id="cb32-30"><a href="#cb32-30" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">length</span>(nm)<span class="sc">&gt;</span> <span class="dv">0</span>){ db <span class="ot">&lt;-</span> db[<span class="sc">-</span>nm, ]}</span>
<span id="cb32-31"><a href="#cb32-31" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(nm)</span>
<span id="cb32-32"><a href="#cb32-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-33"><a href="#cb32-33" aria-hidden="true" tabindex="-1"></a><span class="co"># Separate into interactions with a single and multiple Ensembl Gene IDs</span></span>
<span id="cb32-34"><a href="#cb32-34" aria-hidden="true" tabindex="-1"></a>db_m2m <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">c</span>(<span class="fu">which</span>(<span class="fu">str_detect</span>(<span class="at">string =</span> db<span class="sc">$</span>Ensembl_Gene_Interactor_A, <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">|"</span>) <span class="sc">==</span> <span class="cn">TRUE</span>),</span>
<span id="cb32-35"><a href="#cb32-35" aria-hidden="true" tabindex="-1"></a>                   <span class="fu">which</span>(<span class="fu">str_detect</span>(<span class="at">string =</span> db<span class="sc">$</span>Ensembl_Gene_Interactor_B, <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">|"</span>) <span class="sc">==</span> <span class="cn">TRUE</span>)))</span>
<span id="cb32-36"><a href="#cb32-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-37"><a href="#cb32-37" aria-hidden="true" tabindex="-1"></a>db_o2o <span class="ot">&lt;-</span> <span class="fu">setdiff</span>(<span class="at">x =</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(db), <span class="at">y =</span> db_m2m)</span>
<span id="cb32-38"><a href="#cb32-38" aria-hidden="true" tabindex="-1"></a>db_m2m <span class="ot">&lt;-</span> db[db_m2m, ]</span>
<span id="cb32-39"><a href="#cb32-39" aria-hidden="true" tabindex="-1"></a>db_o2o <span class="ot">&lt;-</span> db[db_o2o, ]</span>
<span id="cb32-40"><a href="#cb32-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-41"><a href="#cb32-41" aria-hidden="true" tabindex="-1"></a><span class="co"># Expand interactions with multiple ensembl gene IDs</span></span>
<span id="cb32-42"><a href="#cb32-42" aria-hidden="true" tabindex="-1"></a>output <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="cn">NA</span>, <span class="at">nrow =</span> <span class="dv">1</span>, <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">dimnames =</span> <span class="fu">list</span>(<span class="st">"ToRemove"</span>, <span class="fu">c</span>(<span class="st">"Ensembl_Gene_Interactor_A"</span>, <span class="st">"Ensembl_Gene_Interactor_B"</span>)))</span>
<span id="cb32-43"><a href="#cb32-43" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(db_m2m))){</span>
<span id="cb32-44"><a href="#cb32-44" aria-hidden="true" tabindex="-1"></a>  in_a <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">str_split</span>(<span class="at">string =</span> db_m2m<span class="sc">$</span>Ensembl_Gene_Interactor_A[i], <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">|</span><span class="sc">\\</span><span class="st">|"</span>))</span>
<span id="cb32-45"><a href="#cb32-45" aria-hidden="true" tabindex="-1"></a>  in_b <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">str_split</span>(<span class="at">string =</span> db_m2m<span class="sc">$</span>Ensembl_Gene_Interactor_B[i], <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">|</span><span class="sc">\\</span><span class="st">|"</span>))</span>
<span id="cb32-46"><a href="#cb32-46" aria-hidden="true" tabindex="-1"></a>  ab <span class="ot">&lt;-</span> <span class="fu">expand.grid</span>(<span class="st">"Ensembl_Gene_Interactor_A"</span> <span class="ot">=</span> in_a, <span class="st">"Ensembl_Gene_Interactor_B"</span> <span class="ot">=</span> in_b)</span>
<span id="cb32-47"><a href="#cb32-47" aria-hidden="true" tabindex="-1"></a>  output <span class="ot">&lt;-</span> <span class="fu">rbind</span>(output, ab)</span>
<span id="cb32-48"><a href="#cb32-48" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb32-49"><a href="#cb32-49" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(in_a, in_b, ab, i)</span>
<span id="cb32-50"><a href="#cb32-50" aria-hidden="true" tabindex="-1"></a>output <span class="ot">&lt;-</span> output[<span class="sc">-</span><span class="dv">1</span>, ]</span>
<span id="cb32-51"><a href="#cb32-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-52"><a href="#cb32-52" aria-hidden="true" tabindex="-1"></a><span class="co"># Combine interactions that map a single and multiple Ensembl Gene IDs </span></span>
<span id="cb32-53"><a href="#cb32-53" aria-hidden="true" tabindex="-1"></a>db_ens <span class="ot">&lt;-</span> <span class="fu">rbind</span>(db_o2o[, <span class="fu">c</span>(<span class="st">"Ensembl_Gene_Interactor_A"</span>, <span class="st">"Ensembl_Gene_Interactor_B"</span>)], </span>
<span id="cb32-54"><a href="#cb32-54" aria-hidden="true" tabindex="-1"></a>                output)</span>
<span id="cb32-55"><a href="#cb32-55" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(db, db_m2m, db_o2o, output)</span>
<span id="cb32-56"><a href="#cb32-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-57"><a href="#cb32-57" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove self interactions</span></span>
<span id="cb32-58"><a href="#cb32-58" aria-hidden="true" tabindex="-1"></a>mark_self <span class="ot">&lt;-</span> <span class="fu">apply</span>(db_ens, <span class="dv">1</span>, <span class="at">FUN =</span> <span class="cf">function</span>(x) <span class="fu">length</span>(<span class="fu">unique</span>(x)))</span>
<span id="cb32-59"><a href="#cb32-59" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">length</span>(<span class="fu">which</span>(mark_self <span class="sc">==</span> <span class="dv">1</span>))<span class="sc">&gt;</span> <span class="dv">0</span>){db_ens <span class="ot">&lt;-</span> db_ens[<span class="sc">-</span><span class="fu">which</span>(mark_self <span class="sc">==</span> <span class="dv">1</span>), ]}</span>
<span id="cb32-60"><a href="#cb32-60" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(mark_self)</span>
<span id="cb32-61"><a href="#cb32-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-62"><a href="#cb32-62" aria-hidden="true" tabindex="-1"></a><span class="co"># Re-order interactions to remove duplicates</span></span>
<span id="cb32-63"><a href="#cb32-63" aria-hidden="true" tabindex="-1"></a>db_ens_order <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">apply</span>(db_ens, <span class="dv">1</span>, <span class="at">FUN =</span> <span class="cf">function</span>(x) x[<span class="fu">order</span>(x)]))</span>
<span id="cb32-64"><a href="#cb32-64" aria-hidden="true" tabindex="-1"></a>db_ens <span class="ot">&lt;-</span> <span class="fu">unique</span>(db_ens_order) </span>
<span id="cb32-65"><a href="#cb32-65" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(db_ens_order)</span>
<span id="cb32-66"><a href="#cb32-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-67"><a href="#cb32-67" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert to data frame</span></span>
<span id="cb32-68"><a href="#cb32-68" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(db_ens) <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(db_ens)</span>
<span id="cb32-69"><a href="#cb32-69" aria-hidden="true" tabindex="-1"></a>db_ens <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(db_ens)</span>
<span id="cb32-70"><a href="#cb32-70" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-71"><a href="#cb32-71" aria-hidden="true" tabindex="-1"></a><span class="co"># Save as tab delimited file </span></span>
<span id="cb32-72"><a href="#cb32-72" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(<span class="at">x =</span> db_ens, </span>
<span id="cb32-73"><a href="#cb32-73" aria-hidden="true" tabindex="-1"></a>            <span class="at">file =</span> <span class="st">"./02_Datasets/MINT/species_human_EnsemblGeneID.txt"</span>, </span>
<span id="cb32-74"><a href="#cb32-74" aria-hidden="true" tabindex="-1"></a>            <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>, <span class="at">col.names =</span> <span class="cn">FALSE</span>, <span class="at">quote =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="reactome" class="level2">
<h2 class="anchored" data-anchor-id="reactome">REACTOME</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Data download</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="at">path =</span> <span class="st">"./02_Datasets/REACTOME/"</span>)</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(<span class="at">url =</span> <span class="st">"https://reactome.org/download/current/interactors/reactome.homo_sapiens.interactions.tab-delimited.txt"</span>,</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>              <span class="at">destfile =</span> <span class="st">"./02_Datasets/REACTOME/reactome.homo_sapiens.interactions.tab-delimited.txt"</span>) </span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a><span class="co"># REACTOME files to system variables </span></span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a><span class="fu">Sys.setenv</span>(<span class="at">REACTOME =</span> <span class="st">"./02_Datasets/REACTOME/reactome.homo_sapiens.interactions.tab-delimited.txt"</span>)</span>
<span id="cb33-9"><a href="#cb33-9" aria-hidden="true" tabindex="-1"></a><span class="fu">Sys.setenv</span>(<span class="at">REACTOME_TMP =</span> <span class="st">"./02_Datasets/REACTOME/reactome.homo_sapiens.interactions.tab-delimited_tmp.txt"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Parse file to remove ' and # from file</span></span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a><span class="fu">sed</span> <span class="st">"s/'//g"</span> <span class="va">$REACTOME</span> <span class="op">&gt;</span> <span class="va">$REACTOME_TMP</span></span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a><span class="fu">sed</span> <span class="st">"s/#//g"</span> <span class="va">$REACTOME_TMP</span> <span class="op">&gt;</span> <span class="va">$REACTOME</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove intermediate files</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a><span class="fu">unlink</span>(<span class="at">x =</span> <span class="st">"./02_Datasets/REACTOME/reactome.homo_sapiens.interactions.tab-delimited_tmp.txt"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Reactome</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>db <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="at">file =</span> <span class="st">"./02_Datasets/REACTOME/reactome.homo_sapiens.interactions.tab-delimited.txt"</span>, <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">quote =</span> <span class="st">""</span>, <span class="at">header =</span> <span class="cn">TRUE</span>)</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Replace "." from colnames </span></span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(db) <span class="ot">&lt;-</span> <span class="fu">str_replace_all</span>(<span class="at">string =</span> <span class="fu">colnames</span>(db), <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">."</span>, <span class="at">replacement =</span> <span class="st">"_"</span>)</span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Mapping through uniprot</span></span>
<span id="cb36-8"><a href="#cb36-8" aria-hidden="true" tabindex="-1"></a>uni <span class="ot">&lt;-</span> <span class="fu">intersect</span>(<span class="fu">which</span>(<span class="fu">str_detect</span>(<span class="at">string =</span> db<span class="sc">$</span>Interactor_1_uniprot_id, <span class="at">pattern =</span> <span class="st">"uniprotkb"</span>) <span class="sc">==</span> <span class="cn">TRUE</span>),</span>
<span id="cb36-9"><a href="#cb36-9" aria-hidden="true" tabindex="-1"></a>                 <span class="fu">which</span>(<span class="fu">str_detect</span>(<span class="at">string =</span> db<span class="sc">$</span>Interactor_2_uniprot_id, <span class="at">pattern =</span> <span class="st">"uniprotkb"</span>) <span class="sc">==</span> <span class="cn">TRUE</span>))</span>
<span id="cb36-10"><a href="#cb36-10" aria-hidden="true" tabindex="-1"></a>db <span class="ot">&lt;-</span> db[uni, ]</span>
<span id="cb36-11"><a href="#cb36-11" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(uni)</span>
<span id="cb36-12"><a href="#cb36-12" aria-hidden="true" tabindex="-1"></a>db<span class="sc">$</span>Interactor_1_uniprot_id <span class="ot">&lt;-</span> <span class="fu">str_replace</span>(<span class="at">string =</span> db<span class="sc">$</span>Interactor_1_uniprot_id, <span class="at">pattern =</span> <span class="st">"uniprotkb:"</span>, <span class="at">replacement =</span> <span class="st">""</span>)</span>
<span id="cb36-13"><a href="#cb36-13" aria-hidden="true" tabindex="-1"></a>db<span class="sc">$</span>Interactor_2_uniprot_id <span class="ot">&lt;-</span> <span class="fu">str_replace</span>(<span class="at">string =</span> db<span class="sc">$</span>Interactor_2_uniprot_id, <span class="at">pattern =</span> <span class="st">"uniprotkb:"</span>, <span class="at">replacement =</span> <span class="st">""</span>)</span>
<span id="cb36-14"><a href="#cb36-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-15"><a href="#cb36-15" aria-hidden="true" tabindex="-1"></a>db <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(db, </span>
<span id="cb36-16"><a href="#cb36-16" aria-hidden="true" tabindex="-1"></a>                 <span class="at">Ensembl_Gene_Interactor_A =</span> uniprot_to_ensembl<span class="sc">$</span>Ensembl_GeneID[<span class="fu">match</span>(<span class="at">x =</span> db<span class="sc">$</span>Interactor_1_uniprot_id, <span class="at">table =</span> uniprot_to_ensembl<span class="sc">$</span>Uniprot_ID)],</span>
<span id="cb36-17"><a href="#cb36-17" aria-hidden="true" tabindex="-1"></a>                 <span class="at">Ensembl_Gene_Interactor_B =</span> uniprot_to_ensembl<span class="sc">$</span>Ensembl_GeneID[<span class="fu">match</span>(<span class="at">x =</span> db<span class="sc">$</span>Interactor_2_uniprot_id, <span class="at">table =</span> uniprot_to_ensembl<span class="sc">$</span>Uniprot_ID)])</span>
<span id="cb36-18"><a href="#cb36-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-19"><a href="#cb36-19" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove interactions where either interactor cannot be mapped to Ensembl Gene IDs</span></span>
<span id="cb36-20"><a href="#cb36-20" aria-hidden="true" tabindex="-1"></a>nm <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">c</span>(<span class="fu">which</span>(<span class="fu">is.na</span>(db<span class="sc">$</span>Ensembl_Gene_Interactor_A) <span class="sc">==</span> <span class="cn">TRUE</span>),</span>
<span id="cb36-21"><a href="#cb36-21" aria-hidden="true" tabindex="-1"></a>               <span class="fu">which</span>(<span class="fu">is.na</span>(db<span class="sc">$</span>Ensembl_Gene_Interactor_B) <span class="sc">==</span> <span class="cn">TRUE</span>)))</span>
<span id="cb36-22"><a href="#cb36-22" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">length</span>(nm)<span class="sc">&gt;</span> <span class="dv">0</span>){ db <span class="ot">&lt;-</span> db[<span class="sc">-</span>nm, ]}</span>
<span id="cb36-23"><a href="#cb36-23" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(nm)</span>
<span id="cb36-24"><a href="#cb36-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-25"><a href="#cb36-25" aria-hidden="true" tabindex="-1"></a><span class="co"># Separate into interactions with a single and multiple Ensembl Gene IDs</span></span>
<span id="cb36-26"><a href="#cb36-26" aria-hidden="true" tabindex="-1"></a>db_m2m <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">c</span>(<span class="fu">which</span>(<span class="fu">str_detect</span>(<span class="at">string =</span> db<span class="sc">$</span>Ensembl_Gene_Interactor_A, <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">|"</span>) <span class="sc">==</span> <span class="cn">TRUE</span>),</span>
<span id="cb36-27"><a href="#cb36-27" aria-hidden="true" tabindex="-1"></a>                   <span class="fu">which</span>(<span class="fu">str_detect</span>(<span class="at">string =</span> db<span class="sc">$</span>Ensembl_Gene_Interactor_B, <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">|"</span>) <span class="sc">==</span> <span class="cn">TRUE</span>)))</span>
<span id="cb36-28"><a href="#cb36-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-29"><a href="#cb36-29" aria-hidden="true" tabindex="-1"></a>db_o2o <span class="ot">&lt;-</span> <span class="fu">setdiff</span>(<span class="at">x =</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(db), <span class="at">y =</span> db_m2m)</span>
<span id="cb36-30"><a href="#cb36-30" aria-hidden="true" tabindex="-1"></a>db_m2m <span class="ot">&lt;-</span> db[db_m2m, ]</span>
<span id="cb36-31"><a href="#cb36-31" aria-hidden="true" tabindex="-1"></a>db_o2o <span class="ot">&lt;-</span> db[db_o2o, ]</span>
<span id="cb36-32"><a href="#cb36-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-33"><a href="#cb36-33" aria-hidden="true" tabindex="-1"></a><span class="co"># Expand interactions with multiple ensembl gene IDs</span></span>
<span id="cb36-34"><a href="#cb36-34" aria-hidden="true" tabindex="-1"></a>output <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="cn">NA</span>, <span class="at">nrow =</span> <span class="dv">1</span>, <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">dimnames =</span> <span class="fu">list</span>(<span class="st">"ToRemove"</span>, <span class="fu">c</span>(<span class="st">"Ensembl_Gene_Interactor_A"</span>, <span class="st">"Ensembl_Gene_Interactor_B"</span>)))</span>
<span id="cb36-35"><a href="#cb36-35" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(db_m2m))){</span>
<span id="cb36-36"><a href="#cb36-36" aria-hidden="true" tabindex="-1"></a>  in_a <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">str_split</span>(<span class="at">string =</span> db_m2m<span class="sc">$</span>Ensembl_Gene_Interactor_A[i], <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">|</span><span class="sc">\\</span><span class="st">|"</span>))</span>
<span id="cb36-37"><a href="#cb36-37" aria-hidden="true" tabindex="-1"></a>  in_b <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">str_split</span>(<span class="at">string =</span> db_m2m<span class="sc">$</span>Ensembl_Gene_Interactor_B[i], <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">|</span><span class="sc">\\</span><span class="st">|"</span>))</span>
<span id="cb36-38"><a href="#cb36-38" aria-hidden="true" tabindex="-1"></a>  ab <span class="ot">&lt;-</span> <span class="fu">expand.grid</span>(<span class="st">"Ensembl_Gene_Interactor_A"</span> <span class="ot">=</span> in_a, <span class="st">"Ensembl_Gene_Interactor_B"</span> <span class="ot">=</span> in_b)</span>
<span id="cb36-39"><a href="#cb36-39" aria-hidden="true" tabindex="-1"></a>  output <span class="ot">&lt;-</span> <span class="fu">rbind</span>(output, ab)</span>
<span id="cb36-40"><a href="#cb36-40" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb36-41"><a href="#cb36-41" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(in_a, in_b, ab, i)</span>
<span id="cb36-42"><a href="#cb36-42" aria-hidden="true" tabindex="-1"></a>output <span class="ot">&lt;-</span> output[<span class="sc">-</span><span class="dv">1</span>, ]</span>
<span id="cb36-43"><a href="#cb36-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-44"><a href="#cb36-44" aria-hidden="true" tabindex="-1"></a><span class="co"># Combine interactions that map a single and multiple Ensembl Gene IDs </span></span>
<span id="cb36-45"><a href="#cb36-45" aria-hidden="true" tabindex="-1"></a>db_ens <span class="ot">&lt;-</span> <span class="fu">rbind</span>(db_o2o[, <span class="fu">c</span>(<span class="st">"Ensembl_Gene_Interactor_A"</span>, <span class="st">"Ensembl_Gene_Interactor_B"</span>)], </span>
<span id="cb36-46"><a href="#cb36-46" aria-hidden="true" tabindex="-1"></a>                output)</span>
<span id="cb36-47"><a href="#cb36-47" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(db, db_m2m, db_o2o, output)</span>
<span id="cb36-48"><a href="#cb36-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-49"><a href="#cb36-49" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove self interactions</span></span>
<span id="cb36-50"><a href="#cb36-50" aria-hidden="true" tabindex="-1"></a>mark_self <span class="ot">&lt;-</span> <span class="fu">apply</span>(db_ens, <span class="dv">1</span>, <span class="at">FUN =</span> <span class="cf">function</span>(x) <span class="fu">length</span>(<span class="fu">unique</span>(x)))</span>
<span id="cb36-51"><a href="#cb36-51" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">length</span>(<span class="fu">which</span>(mark_self <span class="sc">==</span> <span class="dv">1</span>))<span class="sc">&gt;</span> <span class="dv">0</span>){db_ens <span class="ot">&lt;-</span> db_ens[<span class="sc">-</span><span class="fu">which</span>(mark_self <span class="sc">==</span> <span class="dv">1</span>), ]}</span>
<span id="cb36-52"><a href="#cb36-52" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(mark_self)</span>
<span id="cb36-53"><a href="#cb36-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-54"><a href="#cb36-54" aria-hidden="true" tabindex="-1"></a><span class="co"># Re-order interactions to remove duplicates</span></span>
<span id="cb36-55"><a href="#cb36-55" aria-hidden="true" tabindex="-1"></a>db_ens_order <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">apply</span>(db_ens, <span class="dv">1</span>, <span class="at">FUN =</span> <span class="cf">function</span>(x) x[<span class="fu">order</span>(x)]))</span>
<span id="cb36-56"><a href="#cb36-56" aria-hidden="true" tabindex="-1"></a>db_ens <span class="ot">&lt;-</span> <span class="fu">unique</span>(db_ens_order) </span>
<span id="cb36-57"><a href="#cb36-57" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(db_ens_order)</span>
<span id="cb36-58"><a href="#cb36-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-59"><a href="#cb36-59" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert to data frame</span></span>
<span id="cb36-60"><a href="#cb36-60" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(db_ens) <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(db_ens)</span>
<span id="cb36-61"><a href="#cb36-61" aria-hidden="true" tabindex="-1"></a>db_ens <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(db_ens)</span>
<span id="cb36-62"><a href="#cb36-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-63"><a href="#cb36-63" aria-hidden="true" tabindex="-1"></a><span class="co"># Save as tab delimited file </span></span>
<span id="cb36-64"><a href="#cb36-64" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(<span class="at">x =</span> db_ens, </span>
<span id="cb36-65"><a href="#cb36-65" aria-hidden="true" tabindex="-1"></a>            <span class="at">file =</span> <span class="st">"./02_Datasets/REACTOME/reactome.homo_sapiens.interactions.tab-delimited_EnsemblGeneID.txt"</span>, </span>
<span id="cb36-66"><a href="#cb36-66" aria-hidden="true" tabindex="-1"></a>            <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>, <span class="at">col.names =</span> <span class="cn">FALSE</span>, <span class="at">quote =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="string" class="level2">
<h2 class="anchored" data-anchor-id="string">STRING</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Data download</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="at">path =</span> <span class="st">"./02_Datasets/STRING/"</span>)</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(<span class="at">url =</span> <span class="st">"https://stringdb-downloads.org/download/protein.links.v12.0/9606.protein.links.v12.0.txt.gz"</span>,</span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a>              <span class="at">destfile =</span> <span class="st">"./02_Datasets/STRING/9606.protein.links.v12.0.txt.gz"</span>)</span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(<span class="at">url =</span> <span class="st">"https://stringdb-downloads.org/download/protein.physical.links.v12.0/9606.protein.physical.links.v12.0.txt.gz"</span>,</span>
<span id="cb37-8"><a href="#cb37-8" aria-hidden="true" tabindex="-1"></a>              <span class="at">destfile =</span> <span class="st">"./02_Datasets/STRING/9606.protein.physical.links.v12.0.txt.gz"</span>)</span>
<span id="cb37-9"><a href="#cb37-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-10"><a href="#cb37-10" aria-hidden="true" tabindex="-1"></a><span class="fu">gunzip</span>(<span class="st">"./02_Datasets/STRING/9606.protein.links.v12.0.txt.gz"</span>, <span class="at">remove =</span> <span class="cn">TRUE</span>)</span>
<span id="cb37-11"><a href="#cb37-11" aria-hidden="true" tabindex="-1"></a><span class="fu">gunzip</span>(<span class="st">"./02_Datasets/STRING/9606.protein.physical.links.v12.0.txt.gz"</span>, <span class="at">remove =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="protein-links" class="level3">
<h3 class="anchored" data-anchor-id="protein-links">Protein links</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="co"># String</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>db <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="at">file =</span> <span class="st">"./02_Datasets/STRING/9606.protein.links.v12.0.txt"</span>, <span class="at">header =</span> <span class="cn">TRUE</span>)</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>db<span class="sc">$</span>protein1 <span class="ot">&lt;-</span> <span class="fu">str_replace</span>(<span class="at">string =</span> db<span class="sc">$</span>protein1, <span class="at">pattern =</span> <span class="st">"9606."</span>, <span class="at">replacement =</span> <span class="st">""</span>)</span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>db<span class="sc">$</span>protein2 <span class="ot">&lt;-</span> <span class="fu">str_replace</span>(<span class="at">string =</span> db<span class="sc">$</span>protein2, <span class="at">pattern =</span> <span class="st">"9606."</span>, <span class="at">replacement =</span> <span class="st">""</span>)</span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Mapping through protein ID</span></span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a>db <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(db, </span>
<span id="cb38-8"><a href="#cb38-8" aria-hidden="true" tabindex="-1"></a>                 <span class="at">Ensembl_Gene_Interactor_A =</span> proteinID_to_ensembl<span class="sc">$</span>Ensembl_GeneID[<span class="fu">match</span>(<span class="at">x =</span> db<span class="sc">$</span>protein1, <span class="at">table =</span> proteinID_to_ensembl<span class="sc">$</span>Protein_ID)],</span>
<span id="cb38-9"><a href="#cb38-9" aria-hidden="true" tabindex="-1"></a>                 <span class="at">Ensembl_Gene_Interactor_B =</span> proteinID_to_ensembl<span class="sc">$</span>Ensembl_GeneID[<span class="fu">match</span>(<span class="at">x =</span> db<span class="sc">$</span>protein2, <span class="at">table =</span> proteinID_to_ensembl<span class="sc">$</span>Protein_ID)])</span>
<span id="cb38-10"><a href="#cb38-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-11"><a href="#cb38-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove interactions where either interactor cannot be mapped to Ensembl Gene IDs</span></span>
<span id="cb38-12"><a href="#cb38-12" aria-hidden="true" tabindex="-1"></a>nm <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">c</span>(<span class="fu">which</span>(<span class="fu">is.na</span>(db<span class="sc">$</span>Ensembl_Gene_Interactor_A) <span class="sc">==</span> <span class="cn">TRUE</span>),</span>
<span id="cb38-13"><a href="#cb38-13" aria-hidden="true" tabindex="-1"></a>               <span class="fu">which</span>(<span class="fu">is.na</span>(db<span class="sc">$</span>Ensembl_Gene_Interactor_B) <span class="sc">==</span> <span class="cn">TRUE</span>)))</span>
<span id="cb38-14"><a href="#cb38-14" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">length</span>(nm)<span class="sc">&gt;</span> <span class="dv">0</span>){ db <span class="ot">&lt;-</span> db[<span class="sc">-</span>nm, ]}</span>
<span id="cb38-15"><a href="#cb38-15" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(nm)</span>
<span id="cb38-16"><a href="#cb38-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-17"><a href="#cb38-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Check if there are multiple Ensembl Gene IDs</span></span>
<span id="cb38-18"><a href="#cb38-18" aria-hidden="true" tabindex="-1"></a><span class="co">#db_m2m &lt;- unique(c(which(str_detect(string = db$Ensembl_Gene_Interactor_A, pattern = "\\|") == TRUE),</span></span>
<span id="cb38-19"><a href="#cb38-19" aria-hidden="true" tabindex="-1"></a><span class="co">#                   which(str_detect(string = db$Ensembl_Gene_Interactor_B, pattern = "\\|") == TRUE)))</span></span>
<span id="cb38-20"><a href="#cb38-20" aria-hidden="true" tabindex="-1"></a><span class="co">#length(db_m2m)</span></span>
<span id="cb38-21"><a href="#cb38-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-22"><a href="#cb38-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Combine interactions that map a single and multiple Ensembl Gene IDs </span></span>
<span id="cb38-23"><a href="#cb38-23" aria-hidden="true" tabindex="-1"></a>db_ens <span class="ot">&lt;-</span> db[, <span class="fu">c</span>(<span class="st">"Ensembl_Gene_Interactor_A"</span>, <span class="st">"Ensembl_Gene_Interactor_B"</span>)]</span>
<span id="cb38-24"><a href="#cb38-24" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(db)</span>
<span id="cb38-25"><a href="#cb38-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-26"><a href="#cb38-26" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove self interactions</span></span>
<span id="cb38-27"><a href="#cb38-27" aria-hidden="true" tabindex="-1"></a>mark_self <span class="ot">&lt;-</span> <span class="fu">apply</span>(db_ens, <span class="dv">1</span>, <span class="at">FUN =</span> <span class="cf">function</span>(x) <span class="fu">length</span>(<span class="fu">unique</span>(x)))</span>
<span id="cb38-28"><a href="#cb38-28" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">length</span>(<span class="fu">which</span>(mark_self <span class="sc">==</span> <span class="dv">1</span>))<span class="sc">&gt;</span> <span class="dv">0</span>){db_ens <span class="ot">&lt;-</span> db_ens[<span class="sc">-</span><span class="fu">which</span>(mark_self <span class="sc">==</span> <span class="dv">1</span>), ]}</span>
<span id="cb38-29"><a href="#cb38-29" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(mark_self)</span>
<span id="cb38-30"><a href="#cb38-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-31"><a href="#cb38-31" aria-hidden="true" tabindex="-1"></a><span class="co"># Re-order interactions to remove duplicates</span></span>
<span id="cb38-32"><a href="#cb38-32" aria-hidden="true" tabindex="-1"></a>db_ens_order <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">apply</span>(db_ens, <span class="dv">1</span>, <span class="at">FUN =</span> <span class="cf">function</span>(x) x[<span class="fu">order</span>(x)]))</span>
<span id="cb38-33"><a href="#cb38-33" aria-hidden="true" tabindex="-1"></a>db_ens <span class="ot">&lt;-</span> <span class="fu">unique</span>(db_ens_order) </span>
<span id="cb38-34"><a href="#cb38-34" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(db_ens_order)</span>
<span id="cb38-35"><a href="#cb38-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-36"><a href="#cb38-36" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert to data frame</span></span>
<span id="cb38-37"><a href="#cb38-37" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(db_ens) <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(db_ens)</span>
<span id="cb38-38"><a href="#cb38-38" aria-hidden="true" tabindex="-1"></a>db_ens <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(db_ens)</span>
<span id="cb38-39"><a href="#cb38-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-40"><a href="#cb38-40" aria-hidden="true" tabindex="-1"></a><span class="co"># Save as tab delimited file </span></span>
<span id="cb38-41"><a href="#cb38-41" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(<span class="at">x =</span> db_ens, </span>
<span id="cb38-42"><a href="#cb38-42" aria-hidden="true" tabindex="-1"></a>            <span class="at">file =</span> <span class="st">"./02_Datasets/STRING/9606.protein.links.v12.0_EnsemblGeneID.txt"</span>, </span>
<span id="cb38-43"><a href="#cb38-43" aria-hidden="true" tabindex="-1"></a>            <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>, <span class="at">col.names =</span> <span class="cn">FALSE</span>, <span class="at">quote =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="protein-physical" class="level3">
<h3 class="anchored" data-anchor-id="protein-physical">Protein physical</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co"># String</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>db <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="at">file =</span> <span class="st">"./02_Datasets/STRING/9606.protein.physical.links.v12.0.txt"</span>, <span class="at">header =</span> <span class="cn">TRUE</span>)</span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>db<span class="sc">$</span>protein1 <span class="ot">&lt;-</span> <span class="fu">str_replace</span>(<span class="at">string =</span> db<span class="sc">$</span>protein1, <span class="at">pattern =</span> <span class="st">"9606."</span>, <span class="at">replacement =</span> <span class="st">""</span>)</span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>db<span class="sc">$</span>protein2 <span class="ot">&lt;-</span> <span class="fu">str_replace</span>(<span class="at">string =</span> db<span class="sc">$</span>protein2, <span class="at">pattern =</span> <span class="st">"9606."</span>, <span class="at">replacement =</span> <span class="st">""</span>)</span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Mapping through protein ID</span></span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a>db <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(db, </span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true" tabindex="-1"></a>                 <span class="at">Ensembl_Gene_Interactor_A =</span> proteinID_to_ensembl<span class="sc">$</span>Ensembl_GeneID[<span class="fu">match</span>(<span class="at">x =</span> db<span class="sc">$</span>protein1, <span class="at">table =</span> proteinID_to_ensembl<span class="sc">$</span>Protein_ID)],</span>
<span id="cb39-9"><a href="#cb39-9" aria-hidden="true" tabindex="-1"></a>                 <span class="at">Ensembl_Gene_Interactor_B =</span> proteinID_to_ensembl<span class="sc">$</span>Ensembl_GeneID[<span class="fu">match</span>(<span class="at">x =</span> db<span class="sc">$</span>protein2, <span class="at">table =</span> proteinID_to_ensembl<span class="sc">$</span>Protein_ID)])</span>
<span id="cb39-10"><a href="#cb39-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-11"><a href="#cb39-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove interactions where either interactor cannot be mapped to Ensembl Gene IDs</span></span>
<span id="cb39-12"><a href="#cb39-12" aria-hidden="true" tabindex="-1"></a>nm <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">c</span>(<span class="fu">which</span>(<span class="fu">is.na</span>(db<span class="sc">$</span>Ensembl_Gene_Interactor_A) <span class="sc">==</span> <span class="cn">TRUE</span>),</span>
<span id="cb39-13"><a href="#cb39-13" aria-hidden="true" tabindex="-1"></a>               <span class="fu">which</span>(<span class="fu">is.na</span>(db<span class="sc">$</span>Ensembl_Gene_Interactor_B) <span class="sc">==</span> <span class="cn">TRUE</span>)))</span>
<span id="cb39-14"><a href="#cb39-14" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">length</span>(nm)<span class="sc">&gt;</span> <span class="dv">0</span>){ db <span class="ot">&lt;-</span> db[<span class="sc">-</span>nm, ]}</span>
<span id="cb39-15"><a href="#cb39-15" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(nm)</span>
<span id="cb39-16"><a href="#cb39-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-17"><a href="#cb39-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Check if there are multiple Ensembl Gene IDs</span></span>
<span id="cb39-18"><a href="#cb39-18" aria-hidden="true" tabindex="-1"></a><span class="co">#db_m2m &lt;- unique(c(which(str_detect(string = db$Ensembl_Gene_Interactor_A, pattern = "\\|") == TRUE),</span></span>
<span id="cb39-19"><a href="#cb39-19" aria-hidden="true" tabindex="-1"></a><span class="co">#                   which(str_detect(string = db$Ensembl_Gene_Interactor_B, pattern = "\\|") == TRUE)))</span></span>
<span id="cb39-20"><a href="#cb39-20" aria-hidden="true" tabindex="-1"></a><span class="co">#length(db_m2m)</span></span>
<span id="cb39-21"><a href="#cb39-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-22"><a href="#cb39-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Combine interactions that map a single and multiple Ensembl Gene IDs </span></span>
<span id="cb39-23"><a href="#cb39-23" aria-hidden="true" tabindex="-1"></a>db_ens <span class="ot">&lt;-</span> db[, <span class="fu">c</span>(<span class="st">"Ensembl_Gene_Interactor_A"</span>, <span class="st">"Ensembl_Gene_Interactor_B"</span>)]</span>
<span id="cb39-24"><a href="#cb39-24" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(db)</span>
<span id="cb39-25"><a href="#cb39-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-26"><a href="#cb39-26" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove self interactions</span></span>
<span id="cb39-27"><a href="#cb39-27" aria-hidden="true" tabindex="-1"></a>mark_self <span class="ot">&lt;-</span> <span class="fu">apply</span>(db_ens, <span class="dv">1</span>, <span class="at">FUN =</span> <span class="cf">function</span>(x) <span class="fu">length</span>(<span class="fu">unique</span>(x)))</span>
<span id="cb39-28"><a href="#cb39-28" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">length</span>(<span class="fu">which</span>(mark_self <span class="sc">==</span> <span class="dv">1</span>))<span class="sc">&gt;</span> <span class="dv">0</span>){db_ens <span class="ot">&lt;-</span> db_ens[<span class="sc">-</span><span class="fu">which</span>(mark_self <span class="sc">==</span> <span class="dv">1</span>), ]}</span>
<span id="cb39-29"><a href="#cb39-29" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(mark_self)</span>
<span id="cb39-30"><a href="#cb39-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-31"><a href="#cb39-31" aria-hidden="true" tabindex="-1"></a><span class="co"># Re-order interactions to remove duplicates</span></span>
<span id="cb39-32"><a href="#cb39-32" aria-hidden="true" tabindex="-1"></a>db_ens_order <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">apply</span>(db_ens, <span class="dv">1</span>, <span class="at">FUN =</span> <span class="cf">function</span>(x) x[<span class="fu">order</span>(x)]))</span>
<span id="cb39-33"><a href="#cb39-33" aria-hidden="true" tabindex="-1"></a>db_ens <span class="ot">&lt;-</span> <span class="fu">unique</span>(db_ens_order) </span>
<span id="cb39-34"><a href="#cb39-34" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(db_ens_order)</span>
<span id="cb39-35"><a href="#cb39-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-36"><a href="#cb39-36" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert to data frame</span></span>
<span id="cb39-37"><a href="#cb39-37" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(db_ens) <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(db_ens)</span>
<span id="cb39-38"><a href="#cb39-38" aria-hidden="true" tabindex="-1"></a>db_ens <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(db_ens)</span>
<span id="cb39-39"><a href="#cb39-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-40"><a href="#cb39-40" aria-hidden="true" tabindex="-1"></a><span class="co"># Save as tab delimited file </span></span>
<span id="cb39-41"><a href="#cb39-41" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(<span class="at">x =</span> db_ens, </span>
<span id="cb39-42"><a href="#cb39-42" aria-hidden="true" tabindex="-1"></a>            <span class="at">file =</span> <span class="st">"./02_Datasets/STRING/9606.protein.physical.links.v12.0_EnsemblGeneID.txt"</span>, </span>
<span id="cb39-43"><a href="#cb39-43" aria-hidden="true" tabindex="-1"></a>            <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>, <span class="at">col.names =</span> <span class="cn">FALSE</span>, <span class="at">quote =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="overview-resources" class="level2">
<h2 class="anchored" data-anchor-id="overview-resources">Overview resources</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="co"># List all protein protein interactions given in Ensembl Gene IDs</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>fn <span class="ot">&lt;-</span> <span class="fu">list.files</span>(<span class="at">path =</span> <span class="st">"./02_Datasets/"</span>, <span class="at">recursive =</span> <span class="cn">TRUE</span>, <span class="at">pattern =</span> <span class="st">"EnsemblGeneID"</span>, <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>gfn <span class="ot">&lt;-</span> <span class="fu">sapply</span>(<span class="fu">str_split</span>(<span class="at">string =</span> fn, <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">/"</span>), <span class="st">"["</span>, <span class="dv">3</span>)</span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a>gfd <span class="ot">&lt;-</span> <span class="fu">str_replace</span>(<span class="at">string =</span> <span class="fu">sapply</span>(<span class="fu">str_split</span>(<span class="at">string =</span> fn, <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">/"</span>), <span class="st">"["</span>, <span class="dv">4</span>), </span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a>                   <span class="at">pattern =</span> <span class="st">"_EnsemblGeneID.txt"</span>, <span class="at">replacement =</span> <span class="st">""</span>)</span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-8"><a href="#cb40-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Relabel datasets # </span></span>
<span id="cb40-9"><a href="#cb40-9" aria-hidden="true" tabindex="-1"></a>names_datasets <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">fn_sub =</span> <span class="fu">c</span>(<span class="st">"9606_noISI_Q3"</span>, </span>
<span id="cb40-10"><a href="#cb40-10" aria-hidden="true" tabindex="-1"></a>                                        <span class="st">"BIOGRID-ORGANISM-Homo_sapiens-4.4.230.tab3"</span>,</span>
<span id="cb40-11"><a href="#cb40-11" aria-hidden="true" tabindex="-1"></a>                                        <span class="st">"BioPlex_293T_Network_10K_Dec_2019"</span>, <span class="st">"BioPlex_HCT116_Network_5.5K_Dec_2019"</span>,</span>
<span id="cb40-12"><a href="#cb40-12" aria-hidden="true" tabindex="-1"></a>                                        <span class="st">"allComplexes"</span>, <span class="st">"humanComplexes"</span>, <span class="st">"partialComplexes"</span>, </span>
<span id="cb40-13"><a href="#cb40-13" aria-hidden="true" tabindex="-1"></a>                                        <span class="st">"H_sapiens_interactions"</span>,</span>
<span id="cb40-14"><a href="#cb40-14" aria-hidden="true" tabindex="-1"></a>                                        <span class="st">"H-I-05"</span>, <span class="st">"HI-II-14"</span>, <span class="st">"HI-union"</span>, <span class="st">"HuRI"</span>, <span class="st">"Lit-BM"</span>, </span>
<span id="cb40-15"><a href="#cb40-15" aria-hidden="true" tabindex="-1"></a>                                        <span class="st">"Test_space_screens-19"</span>, <span class="st">"Venkatesan-09"</span>, <span class="st">"Yang-16"</span>, <span class="st">"Yu-11"</span>,</span>
<span id="cb40-16"><a href="#cb40-16" aria-hidden="true" tabindex="-1"></a>                                        <span class="st">"intact"</span>, </span>
<span id="cb40-17"><a href="#cb40-17" aria-hidden="true" tabindex="-1"></a>                                        <span class="st">"species_human"</span>,</span>
<span id="cb40-18"><a href="#cb40-18" aria-hidden="true" tabindex="-1"></a>                                        <span class="st">"reactome.homo_sapiens.interactions.tab-delimited"</span>,</span>
<span id="cb40-19"><a href="#cb40-19" aria-hidden="true" tabindex="-1"></a>                                        <span class="st">"9606.protein.links.v12.0"</span>, <span class="st">"9606.protein.physical.links.v12.0"</span>),</span>
<span id="cb40-20"><a href="#cb40-20" aria-hidden="true" tabindex="-1"></a>                             <span class="at">Dataset =</span> <span class="fu">c</span>(<span class="st">"APID"</span>, </span>
<span id="cb40-21"><a href="#cb40-21" aria-hidden="true" tabindex="-1"></a>                                         <span class="st">"BIOGRID"</span>, </span>
<span id="cb40-22"><a href="#cb40-22" aria-hidden="true" tabindex="-1"></a>                                         <span class="st">"BIOPLEX_293T"</span>, <span class="st">"BIOPLEX_HCT116"</span>,</span>
<span id="cb40-23"><a href="#cb40-23" aria-hidden="true" tabindex="-1"></a>                                         <span class="st">"CORUM_All"</span>, <span class="st">"CORUM_human"</span>, <span class="st">"CORUM_partial"</span>,  </span>
<span id="cb40-24"><a href="#cb40-24" aria-hidden="true" tabindex="-1"></a>                                         <span class="st">"HITPREDICT_latest"</span>,</span>
<span id="cb40-25"><a href="#cb40-25" aria-hidden="true" tabindex="-1"></a>                                         <span class="st">"HURI_H-I-05"</span>, <span class="st">"HURI_HI-II-14"</span>, <span class="st">"HURI_HI-union"</span>, <span class="st">"HURI_HuRI"</span>, <span class="st">"HURI_Lit-BM"</span>,</span>
<span id="cb40-26"><a href="#cb40-26" aria-hidden="true" tabindex="-1"></a>                                         <span class="st">"HURI_Test19"</span>, <span class="st">"HURI_Venkatesan"</span>, <span class="st">"HURI_Yang"</span>, <span class="st">"HURI_Yu"</span>,</span>
<span id="cb40-27"><a href="#cb40-27" aria-hidden="true" tabindex="-1"></a>                                         <span class="st">"INTACT"</span>, </span>
<span id="cb40-28"><a href="#cb40-28" aria-hidden="true" tabindex="-1"></a>                                         <span class="st">"MINT"</span>, </span>
<span id="cb40-29"><a href="#cb40-29" aria-hidden="true" tabindex="-1"></a>                                         <span class="st">"REACTOME"</span>,</span>
<span id="cb40-30"><a href="#cb40-30" aria-hidden="true" tabindex="-1"></a>                                         <span class="st">"STRING_Links"</span>, <span class="st">"STRING_Physical"</span>))</span>
<span id="cb40-31"><a href="#cb40-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-32"><a href="#cb40-32" aria-hidden="true" tabindex="-1"></a><span class="co"># Obtain an overview of each</span></span>
<span id="cb40-33"><a href="#cb40-33" aria-hidden="true" tabindex="-1"></a>output <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="cn">NA</span>, <span class="fu">length</span>(fn), <span class="dv">4</span>, </span>
<span id="cb40-34"><a href="#cb40-34" aria-hidden="true" tabindex="-1"></a>                 <span class="at">dimnames =</span> <span class="fu">list</span>(gfd, <span class="fu">c</span>(<span class="st">"Genes"</span>, <span class="st">"Interactions"</span>, <span class="st">"Mean per gene"</span>, <span class="st">"Median per gene"</span>)))</span>
<span id="cb40-35"><a href="#cb40-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-36"><a href="#cb40-36" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(fn))){ </span>
<span id="cb40-37"><a href="#cb40-37" aria-hidden="true" tabindex="-1"></a>  fni <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="at">file =</span> fn[i], <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>)</span>
<span id="cb40-38"><a href="#cb40-38" aria-hidden="true" tabindex="-1"></a>  output[i, <span class="dv">1</span>] <span class="ot">&lt;-</span> <span class="fu">length</span>(<span class="fu">unique</span>(<span class="fu">c</span>(fni<span class="sc">$</span>V1, fni<span class="sc">$</span>V2)))</span>
<span id="cb40-39"><a href="#cb40-39" aria-hidden="true" tabindex="-1"></a>  output[i, <span class="dv">2</span>] <span class="ot">&lt;-</span> <span class="fu">nrow</span>(fni)</span>
<span id="cb40-40"><a href="#cb40-40" aria-hidden="true" tabindex="-1"></a>  output[i, <span class="dv">3</span>] <span class="ot">&lt;-</span> <span class="fu">mean</span>(<span class="fu">table</span>(<span class="fu">c</span>(fni<span class="sc">$</span>V1, fni<span class="sc">$</span>V2)))</span>
<span id="cb40-41"><a href="#cb40-41" aria-hidden="true" tabindex="-1"></a>  output[i, <span class="dv">4</span>] <span class="ot">&lt;-</span> <span class="fu">median</span>(<span class="fu">table</span>(<span class="fu">c</span>(fni<span class="sc">$</span>V1, fni<span class="sc">$</span>V2)))</span>
<span id="cb40-42"><a href="#cb40-42" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb40-43"><a href="#cb40-43" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(fni, i)</span>
<span id="cb40-44"><a href="#cb40-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-45"><a href="#cb40-45" aria-hidden="true" tabindex="-1"></a>output <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">Source =</span> gfn, <span class="at">fn_sub =</span> gfd, output)</span>
<span id="cb40-46"><a href="#cb40-46" aria-hidden="true" tabindex="-1"></a>output <span class="ot">&lt;-</span> <span class="fu">merge</span>(output, names_datasets, <span class="at">by =</span> <span class="st">"fn_sub"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Number of genes</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>g1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data =</span> output, <span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">x =</span> Dataset, <span class="at">y =</span> Genes, <span class="at">fill =</span> Source, <span class="at">label =</span> Genes)) <span class="sc">+</span> </span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_col</span>() <span class="sc">+</span> <span class="fu">geom_text</span>(<span class="at">angle =</span> <span class="dv">90</span>, <span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span> <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>, </span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">90</span>, <span class="at">hjust =</span> <span class="dv">1</span>, <span class="at">vjust =</span> <span class="dv">0</span>))</span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Number of interactions</span></span>
<span id="cb41-8"><a href="#cb41-8" aria-hidden="true" tabindex="-1"></a>g2 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data =</span> output, <span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">x =</span> Dataset, <span class="at">y =</span> Interactions, </span>
<span id="cb41-9"><a href="#cb41-9" aria-hidden="true" tabindex="-1"></a>                                          <span class="at">fill =</span> Source, <span class="at">label =</span> Interactions)) <span class="sc">+</span> </span>
<span id="cb41-10"><a href="#cb41-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_col</span>() <span class="sc">+</span> <span class="fu">geom_text</span>(<span class="at">angle =</span> <span class="dv">90</span>, <span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span> <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb41-11"><a href="#cb41-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>, </span>
<span id="cb41-12"><a href="#cb41-12" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">90</span>, <span class="at">hjust =</span> <span class="dv">1</span>, <span class="at">vjust =</span> <span class="dv">0</span>))</span>
<span id="cb41-13"><a href="#cb41-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-14"><a href="#cb41-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Mean interactions per gene</span></span>
<span id="cb41-15"><a href="#cb41-15" aria-hidden="true" tabindex="-1"></a>g3 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data =</span> output, <span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">x =</span> Dataset, <span class="at">y =</span> Mean.per.gene, </span>
<span id="cb41-16"><a href="#cb41-16" aria-hidden="true" tabindex="-1"></a>                                          <span class="at">fill =</span> Source)) <span class="sc">+</span> </span>
<span id="cb41-17"><a href="#cb41-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_col</span>() <span class="sc">+</span> <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb41-18"><a href="#cb41-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>, </span>
<span id="cb41-19"><a href="#cb41-19" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">90</span>, <span class="at">hjust =</span> <span class="dv">1</span>, <span class="at">vjust =</span> <span class="dv">0</span>))</span>
<span id="cb41-20"><a href="#cb41-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-21"><a href="#cb41-21" aria-hidden="true" tabindex="-1"></a><span class="co"># Median interactions per gene</span></span>
<span id="cb41-22"><a href="#cb41-22" aria-hidden="true" tabindex="-1"></a>g4 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data =</span> output, <span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">x =</span> Dataset, <span class="at">y =</span> Median.per.gene, </span>
<span id="cb41-23"><a href="#cb41-23" aria-hidden="true" tabindex="-1"></a>                                          <span class="at">fill =</span> Source)) <span class="sc">+</span> </span>
<span id="cb41-24"><a href="#cb41-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_col</span>() <span class="sc">+</span> <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb41-25"><a href="#cb41-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>, </span>
<span id="cb41-26"><a href="#cb41-26" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">90</span>, <span class="at">hjust =</span> <span class="dv">1</span>, <span class="at">vjust =</span> <span class="dv">0</span>))</span>
<span id="cb41-27"><a href="#cb41-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-28"><a href="#cb41-28" aria-hidden="true" tabindex="-1"></a><span class="co"># Overview</span></span>
<span id="cb41-29"><a href="#cb41-29" aria-hidden="true" tabindex="-1"></a>g5 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data =</span> output, <span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">log10</span>(Interactions), </span>
<span id="cb41-30"><a href="#cb41-30" aria-hidden="true" tabindex="-1"></a>                                          <span class="at">y =</span> Genes, </span>
<span id="cb41-31"><a href="#cb41-31" aria-hidden="true" tabindex="-1"></a>                                          <span class="at">color =</span> Source, </span>
<span id="cb41-32"><a href="#cb41-32" aria-hidden="true" tabindex="-1"></a>                                          <span class="at">size =</span> Median.per.gene,</span>
<span id="cb41-33"><a href="#cb41-33" aria-hidden="true" tabindex="-1"></a>                                          <span class="at">label =</span> Dataset)) <span class="sc">+</span> </span>
<span id="cb41-34"><a href="#cb41-34" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span> </span>
<span id="cb41-35"><a href="#cb41-35" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text_repel</span>() <span class="sc">+</span> </span>
<span id="cb41-36"><a href="#cb41-36" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span> </span>
<span id="cb41-37"><a href="#cb41-37" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_met_d</span>(<span class="at">name =</span> <span class="st">"Signac"</span>)</span>
<span id="cb41-38"><a href="#cb41-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-39"><a href="#cb41-39" aria-hidden="true" tabindex="-1"></a>g1</span>
<span id="cb41-40"><a href="#cb41-40" aria-hidden="true" tabindex="-1"></a>g2</span>
<span id="cb41-41"><a href="#cb41-41" aria-hidden="true" tabindex="-1"></a>g3</span>
<span id="cb41-42"><a href="#cb41-42" aria-hidden="true" tabindex="-1"></a>g4</span>
<span id="cb41-43"><a href="#cb41-43" aria-hidden="true" tabindex="-1"></a>g5</span>
<span id="cb41-44"><a href="#cb41-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-45"><a href="#cb41-45" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(g1, g2, g3, g4, g5, output)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="merged-ppi-network" class="level2">
<h2 class="anchored" data-anchor-id="merged-ppi-network">Merged PPI network</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create directory for merged PPI</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="at">path =</span> <span class="st">"./03_PPIMerged/"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="all" class="level3">
<h3 class="anchored" data-anchor-id="all">All</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Combine all datasets</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>PPIall <span class="ot">&lt;-</span> <span class="fu">read.table</span>(fn[<span class="dv">1</span>], <span class="at">header =</span> <span class="cn">FALSE</span>)</span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="fu">c</span>(<span class="dv">2</span><span class="sc">:</span><span class="fu">length</span>(fn))){</span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a>  ppi <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="at">file =</span> fn[i], <span class="at">header =</span> <span class="cn">FALSE</span>)</span>
<span id="cb43-6"><a href="#cb43-6" aria-hidden="true" tabindex="-1"></a>  PPIall <span class="ot">&lt;-</span> <span class="fu">rbind</span>(PPIall, ppi)</span>
<span id="cb43-7"><a href="#cb43-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rm</span>(ppi)</span>
<span id="cb43-8"><a href="#cb43-8" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb43-9"><a href="#cb43-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-10"><a href="#cb43-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Keep only unique interactions</span></span>
<span id="cb43-11"><a href="#cb43-11" aria-hidden="true" tabindex="-1"></a>PPIall <span class="ot">&lt;-</span> <span class="fu">unique</span>(PPIall)</span>
<span id="cb43-12"><a href="#cb43-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-13"><a href="#cb43-13" aria-hidden="true" tabindex="-1"></a>fn_ppi <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="fu">c</span>(<span class="st">"./03_PPIMerged/"</span>,</span>
<span id="cb43-14"><a href="#cb43-14" aria-hidden="true" tabindex="-1"></a>                   <span class="st">"MergedPPI_"</span>,</span>
<span id="cb43-15"><a href="#cb43-15" aria-hidden="true" tabindex="-1"></a>                   <span class="fu">paste0</span>(<span class="fu">sapply</span>(<span class="fu">str_split</span>(<span class="fu">date</span>(), <span class="st">" "</span>), <span class="st">"["</span>, <span class="fu">c</span>(<span class="dv">5</span>, <span class="dv">2</span>)), <span class="at">collapse =</span> <span class="st">""</span>),</span>
<span id="cb43-16"><a href="#cb43-16" aria-hidden="true" tabindex="-1"></a>                   <span class="co"># unique(gfn), # Optional to include names of datasets in output file</span></span>
<span id="cb43-17"><a href="#cb43-17" aria-hidden="true" tabindex="-1"></a>                   <span class="st">".txt"</span>),</span>
<span id="cb43-18"><a href="#cb43-18" aria-hidden="true" tabindex="-1"></a>                 <span class="at">collapse =</span> <span class="st">""</span>)</span>
<span id="cb43-19"><a href="#cb43-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-20"><a href="#cb43-20" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(<span class="at">x =</span> PPIall, <span class="at">file =</span> fn_ppi, <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, </span>
<span id="cb43-21"><a href="#cb43-21" aria-hidden="true" tabindex="-1"></a>            <span class="at">quote =</span> <span class="cn">FALSE</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>, <span class="at">col.names =</span> <span class="cn">FALSE</span>)</span>
<span id="cb43-22"><a href="#cb43-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-23"><a href="#cb43-23" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(PPIall)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Combine datasets except string links</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>fn <span class="ot">&lt;-</span> <span class="fu">setdiff</span>(fn, <span class="at">y =</span> <span class="st">"./02_Datasets/STRING/9606.protein.links.v12.0_EnsemblGeneID.txt"</span>)</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>PPIall <span class="ot">&lt;-</span> <span class="fu">read.table</span>(fn[<span class="dv">1</span>], <span class="at">header =</span> <span class="cn">FALSE</span>)</span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="fu">c</span>(<span class="dv">2</span><span class="sc">:</span><span class="fu">length</span>(fn))){</span>
<span id="cb44-7"><a href="#cb44-7" aria-hidden="true" tabindex="-1"></a>  ppi <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="at">file =</span> fn[i], <span class="at">header =</span> <span class="cn">FALSE</span>)</span>
<span id="cb44-8"><a href="#cb44-8" aria-hidden="true" tabindex="-1"></a>  PPIall <span class="ot">&lt;-</span> <span class="fu">rbind</span>(PPIall, ppi)</span>
<span id="cb44-9"><a href="#cb44-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rm</span>(ppi)</span>
<span id="cb44-10"><a href="#cb44-10" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb44-11"><a href="#cb44-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-12"><a href="#cb44-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Keep only unique interactions</span></span>
<span id="cb44-13"><a href="#cb44-13" aria-hidden="true" tabindex="-1"></a>PPIall <span class="ot">&lt;-</span> <span class="fu">unique</span>(PPIall)</span>
<span id="cb44-14"><a href="#cb44-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-15"><a href="#cb44-15" aria-hidden="true" tabindex="-1"></a>fn_ppi <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="fu">c</span>(<span class="st">"./03_PPIMerged/"</span>,</span>
<span id="cb44-16"><a href="#cb44-16" aria-hidden="true" tabindex="-1"></a>                   <span class="st">"MergedPPI_"</span>,</span>
<span id="cb44-17"><a href="#cb44-17" aria-hidden="true" tabindex="-1"></a>                   <span class="fu">paste0</span>(<span class="fu">sapply</span>(<span class="fu">str_split</span>(<span class="fu">date</span>(), <span class="st">" "</span>), <span class="st">"["</span>, <span class="fu">c</span>(<span class="dv">5</span>, <span class="dv">2</span>)), <span class="at">collapse =</span> <span class="st">""</span>),</span>
<span id="cb44-18"><a href="#cb44-18" aria-hidden="true" tabindex="-1"></a>                   <span class="co"># unique(gfn), # Optional to include names of datasets in output file</span></span>
<span id="cb44-19"><a href="#cb44-19" aria-hidden="true" tabindex="-1"></a>                   <span class="st">"_PIO_STRING.txt"</span>),</span>
<span id="cb44-20"><a href="#cb44-20" aria-hidden="true" tabindex="-1"></a>                 <span class="at">collapse =</span> <span class="st">""</span>)</span>
<span id="cb44-21"><a href="#cb44-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-22"><a href="#cb44-22" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(<span class="at">x =</span> PPIall, <span class="at">file =</span> fn_ppi, <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, </span>
<span id="cb44-23"><a href="#cb44-23" aria-hidden="true" tabindex="-1"></a>            <span class="at">quote =</span> <span class="cn">FALSE</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>, <span class="at">col.names =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="references" class="level2">
<h2 class="anchored" data-anchor-id="references">References</h2>
<ol type="1">
<li><p><a href="https://pubmed.ncbi.nlm.nih.gov/30715274/"><em>APID</em> Alonso-Lopez, D., et al.&nbsp;(2019). “APID database: redefining protein-protein interaction experimental evidences and binary interactomes.” Database (Oxford) 2019.</a></p></li>
<li><p><a href="https://pubmed.ncbi.nlm.nih.gov/27131791/"><em>APID</em> Alonso-Lopez, D., et al.&nbsp;(2016). “APID interactomes: providing proteome-based interactomes with controlled quality for multiple species and derived networks.” Nucleic Acids Res 44(W1): W529-535.</a></p></li>
<li><p><a href="https://pubmed.ncbi.nlm.nih.gov/33070389/"><em>BIOGRID</em> Oughtred, R., et al.&nbsp;(2021). “The BioGRID database: A comprehensive biomedical resource of curated protein, genetic, and chemical interactions.” Protein Sci 30(1): 187-200.</a></p></li>
<li><p><a href="https://pubmed.ncbi.nlm.nih.gov/33961781/"><em>BIOPLEX</em> Huttlin, E. L., et al.&nbsp;(2021). “Dual proteome-scale networks reveal cell-specific remodeling of the human interactome.” Cell 184(11): 3022-3040 e3028.</a></p></li>
<li><p><a href="https://pubmed.ncbi.nlm.nih.gov/36382402/"><em>CORUM</em> Tsitsiridis, G., et al.&nbsp;(2023). “CORUM: the comprehensive resource of mammalian protein complexes-2022.” Nucleic Acids Res 51(D1): D539-D545.</a></p></li>
<li><p><a href="https://pubmed.ncbi.nlm.nih.gov/30357367/"><em>CORUM</em> Giurgiu, M., et al.&nbsp;(2019). “CORUM: the comprehensive resource of mammalian protein complexes-2019.” Nucleic Acids Res 47(D1): D559-D563.</a></p></li>
<li><p><a href="https://pubmed.ncbi.nlm.nih.gov/26708988/"><em>HITPREDICT</em> Lopez, Y., et al.&nbsp;(2015). “HitPredict version 4: comprehensive reliability scoring of physical protein-protein interactions from more than 100 species.” Database (Oxford) 2015.</a></p></li>
<li><p><a href="https://pubmed.ncbi.nlm.nih.gov/32296183/"><em>HURI</em> Luck, K., et al.&nbsp;(2020). “A reference map of the human binary protein interactome.” Nature 580(7803): 402-408.</a></p></li>
<li><p><a href="https://pubmed.ncbi.nlm.nih.gov/24234451/"><em>INTACT</em> Orchard, S., et al.&nbsp;(2014). “The MIntAct project–IntAct as a common curation platform for 11 molecular interaction databases.” Nucleic Acids Res 42(Database issue): D358-363.</a></p></li>
<li><p><a href="https://pubmed.ncbi.nlm.nih.gov/22096227/"><em>MINT</em> Licata, L., et al.&nbsp;(2012). “MINT, the molecular interaction database: 2012 update.” Nucleic Acids Res 40(Database issue): D857-861.</a></p></li>
<li><p><a href="https://pubmed.ncbi.nlm.nih.gov/34788843/"><em>REACTOME</em> Gillespie, M., et al.&nbsp;(2022). “The reactome pathway knowledgebase 2022.” Nucleic Acids Res 50(D1): D687-D692.</a></p></li>
<li><p><a href="https://pubmed.ncbi.nlm.nih.gov/36370105/"><em>STRING</em> Szklarczyk, D., et al.&nbsp;(2023). “The STRING database in 2023: protein-protein association networks and functional enrichment analyses for any sequenced genome of interest.” Nucleic Acids Res 51(D1): D638-D646.</a></p></li>
</ol>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>